{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Data input\n",
    "=========="
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0m\u001b[01;32mall.BRCA-EU.mutations-context.tsv\u001b[0m*\r\n",
      "\u001b[01;34meach_gene_in_BRCA-EU\u001b[0m/\r\n",
      "\u001b[01;34mmain-genes_BRCA-EU\u001b[0m/\r\n"
     ]
    }
   ],
   "source": [
    "%ls ~/results/ | grep BRCA-EU"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function read_table in module pandas.io.parsers:\n",
      "\n",
      "read_table(filepath_or_buffer, sep='\\t', delimiter=None, header='infer', names=None, index_col=None, usecols=None, squeeze=False, prefix=None, mangle_dupe_cols=True, dtype=None, engine=None, converters=None, true_values=None, false_values=None, skipinitialspace=False, skiprows=None, nrows=None, na_values=None, keep_default_na=True, na_filter=True, verbose=False, skip_blank_lines=True, parse_dates=False, infer_datetime_format=False, keep_date_col=False, date_parser=None, dayfirst=False, iterator=False, chunksize=None, compression='infer', thousands=None, decimal=b'.', lineterminator=None, quotechar='\"', quoting=0, escapechar=None, comment=None, encoding=None, dialect=None, tupleize_cols=None, error_bad_lines=True, warn_bad_lines=True, skipfooter=0, skip_footer=0, doublequote=True, delim_whitespace=False, as_recarray=None, compact_ints=None, use_unsigned=None, low_memory=True, buffer_lines=None, memory_map=False, float_precision=None)\n",
      "    Read general delimited file into DataFrame\n",
      "    \n",
      "    Also supports optionally iterating or breaking of the file\n",
      "    into chunks.\n",
      "    \n",
      "    Additional help can be found in the `online docs for IO Tools\n",
      "    <http://pandas.pydata.org/pandas-docs/stable/io.html>`_.\n",
      "    \n",
      "    Parameters\n",
      "    ----------\n",
      "    filepath_or_buffer : str, pathlib.Path, py._path.local.LocalPath or any object with a read() method (such as a file handle or StringIO)\n",
      "        The string could be a URL. Valid URL schemes include http, ftp, s3, and\n",
      "        file. For file URLs, a host is expected. For instance, a local file could\n",
      "        be file ://localhost/path/to/table.csv\n",
      "    sep : str, default \\t (tab-stop)\n",
      "        Delimiter to use. If sep is None, the C engine cannot automatically detect\n",
      "        the separator, but the Python parsing engine can, meaning the latter will\n",
      "        be used and automatically detect the separator by Python's builtin sniffer\n",
      "        tool, ``csv.Sniffer``. In addition, separators longer than 1 character and\n",
      "        different from ``'\\s+'`` will be interpreted as regular expressions and\n",
      "        will also force the use of the Python parsing engine. Note that regex\n",
      "        delimiters are prone to ignoring quoted data. Regex example: ``'\\r\\t'``\n",
      "    delimiter : str, default ``None``\n",
      "        Alternative argument name for sep.\n",
      "    delim_whitespace : boolean, default False\n",
      "        Specifies whether or not whitespace (e.g. ``' '`` or ``'    '``) will be\n",
      "        used as the sep. Equivalent to setting ``sep='\\s+'``. If this option\n",
      "        is set to True, nothing should be passed in for the ``delimiter``\n",
      "        parameter.\n",
      "    \n",
      "        .. versionadded:: 0.18.1 support for the Python parser.\n",
      "    \n",
      "    header : int or list of ints, default 'infer'\n",
      "        Row number(s) to use as the column names, and the start of the data.\n",
      "        Default behavior is as if set to 0 if no ``names`` passed, otherwise\n",
      "        ``None``. Explicitly pass ``header=0`` to be able to replace existing\n",
      "        names. The header can be a list of integers that specify row locations for\n",
      "        a multi-index on the columns e.g. [0,1,3]. Intervening rows that are not\n",
      "        specified will be skipped (e.g. 2 in this example is skipped). Note that\n",
      "        this parameter ignores commented lines and empty lines if\n",
      "        ``skip_blank_lines=True``, so header=0 denotes the first line of data\n",
      "        rather than the first line of the file.\n",
      "    names : array-like, default None\n",
      "        List of column names to use. If file contains no header row, then you\n",
      "        should explicitly pass header=None. Duplicates in this list will cause\n",
      "        a ``UserWarning`` to be issued.\n",
      "    index_col : int or sequence or False, default None\n",
      "        Column to use as the row labels of the DataFrame. If a sequence is given, a\n",
      "        MultiIndex is used. If you have a malformed file with delimiters at the end\n",
      "        of each line, you might consider index_col=False to force pandas to _not_\n",
      "        use the first column as the index (row names)\n",
      "    usecols : array-like or callable, default None\n",
      "        Return a subset of the columns. If array-like, all elements must either\n",
      "        be positional (i.e. integer indices into the document columns) or strings\n",
      "        that correspond to column names provided either by the user in `names` or\n",
      "        inferred from the document header row(s). For example, a valid array-like\n",
      "        `usecols` parameter would be [0, 1, 2] or ['foo', 'bar', 'baz'].\n",
      "    \n",
      "        If callable, the callable function will be evaluated against the column\n",
      "        names, returning names where the callable function evaluates to True. An\n",
      "        example of a valid callable argument would be ``lambda x: x.upper() in\n",
      "        ['AAA', 'BBB', 'DDD']``. Using this parameter results in much faster\n",
      "        parsing time and lower memory usage.\n",
      "    as_recarray : boolean, default False\n",
      "        .. deprecated:: 0.19.0\n",
      "           Please call `pd.read_csv(...).to_records()` instead.\n",
      "    \n",
      "        Return a NumPy recarray instead of a DataFrame after parsing the data.\n",
      "        If set to True, this option takes precedence over the `squeeze` parameter.\n",
      "        In addition, as row indices are not available in such a format, the\n",
      "        `index_col` parameter will be ignored.\n",
      "    squeeze : boolean, default False\n",
      "        If the parsed data only contains one column then return a Series\n",
      "    prefix : str, default None\n",
      "        Prefix to add to column numbers when no header, e.g. 'X' for X0, X1, ...\n",
      "    mangle_dupe_cols : boolean, default True\n",
      "        Duplicate columns will be specified as 'X.0'...'X.N', rather than\n",
      "        'X'...'X'. Passing in False will cause data to be overwritten if there\n",
      "        are duplicate names in the columns.\n",
      "    dtype : Type name or dict of column -> type, default None\n",
      "        Data type for data or columns. E.g. {'a': np.float64, 'b': np.int32}\n",
      "        Use `str` or `object` to preserve and not interpret dtype.\n",
      "        If converters are specified, they will be applied INSTEAD\n",
      "        of dtype conversion.\n",
      "    engine : {'c', 'python'}, optional\n",
      "        Parser engine to use. The C engine is faster while the python engine is\n",
      "        currently more feature-complete.\n",
      "    converters : dict, default None\n",
      "        Dict of functions for converting values in certain columns. Keys can either\n",
      "        be integers or column labels\n",
      "    true_values : list, default None\n",
      "        Values to consider as True\n",
      "    false_values : list, default None\n",
      "        Values to consider as False\n",
      "    skipinitialspace : boolean, default False\n",
      "        Skip spaces after delimiter.\n",
      "    skiprows : list-like or integer or callable, default None\n",
      "        Line numbers to skip (0-indexed) or number of lines to skip (int)\n",
      "        at the start of the file.\n",
      "    \n",
      "        If callable, the callable function will be evaluated against the row\n",
      "        indices, returning True if the row should be skipped and False otherwise.\n",
      "        An example of a valid callable argument would be ``lambda x: x in [0, 2]``.\n",
      "    skipfooter : int, default 0\n",
      "        Number of lines at bottom of file to skip (Unsupported with engine='c')\n",
      "    skip_footer : int, default 0\n",
      "        .. deprecated:: 0.19.0\n",
      "           Use the `skipfooter` parameter instead, as they are identical\n",
      "    nrows : int, default None\n",
      "        Number of rows of file to read. Useful for reading pieces of large files\n",
      "    na_values : scalar, str, list-like, or dict, default None\n",
      "        Additional strings to recognize as NA/NaN. If dict passed, specific\n",
      "        per-column NA values.  By default the following values are interpreted as\n",
      "        NaN: '', '#N/A', '#N/A N/A', '#NA', '-1.#IND', '-1.#QNAN', '-NaN', '-nan',\n",
      "        '1.#IND', '1.#QNAN', 'N/A', 'NA', 'NULL', 'NaN', 'n/a', 'nan',\n",
      "        'null'.\n",
      "    keep_default_na : bool, default True\n",
      "        If na_values are specified and keep_default_na is False the default NaN\n",
      "        values are overridden, otherwise they're appended to.\n",
      "    na_filter : boolean, default True\n",
      "        Detect missing value markers (empty strings and the value of na_values). In\n",
      "        data without any NAs, passing na_filter=False can improve the performance\n",
      "        of reading a large file\n",
      "    verbose : boolean, default False\n",
      "        Indicate number of NA values placed in non-numeric columns\n",
      "    skip_blank_lines : boolean, default True\n",
      "        If True, skip over blank lines rather than interpreting as NaN values\n",
      "    parse_dates : boolean or list of ints or names or list of lists or dict, default False\n",
      "    \n",
      "        * boolean. If True -> try parsing the index.\n",
      "        * list of ints or names. e.g. If [1, 2, 3] -> try parsing columns 1, 2, 3\n",
      "          each as a separate date column.\n",
      "        * list of lists. e.g.  If [[1, 3]] -> combine columns 1 and 3 and parse as\n",
      "          a single date column.\n",
      "        * dict, e.g. {'foo' : [1, 3]} -> parse columns 1, 3 as date and call result\n",
      "          'foo'\n",
      "    \n",
      "        If a column or index contains an unparseable date, the entire column or\n",
      "        index will be returned unaltered as an object data type. For non-standard\n",
      "        datetime parsing, use ``pd.to_datetime`` after ``pd.read_csv``\n",
      "    \n",
      "        Note: A fast-path exists for iso8601-formatted dates.\n",
      "    infer_datetime_format : boolean, default False\n",
      "        If True and `parse_dates` is enabled, pandas will attempt to infer the\n",
      "        format of the datetime strings in the columns, and if it can be inferred,\n",
      "        switch to a faster method of parsing them. In some cases this can increase\n",
      "        the parsing speed by 5-10x.\n",
      "    keep_date_col : boolean, default False\n",
      "        If True and `parse_dates` specifies combining multiple columns then\n",
      "        keep the original columns.\n",
      "    date_parser : function, default None\n",
      "        Function to use for converting a sequence of string columns to an array of\n",
      "        datetime instances. The default uses ``dateutil.parser.parser`` to do the\n",
      "        conversion. Pandas will try to call `date_parser` in three different ways,\n",
      "        advancing to the next if an exception occurs: 1) Pass one or more arrays\n",
      "        (as defined by `parse_dates`) as arguments; 2) concatenate (row-wise) the\n",
      "        string values from the columns defined by `parse_dates` into a single array\n",
      "        and pass that; and 3) call `date_parser` once for each row using one or\n",
      "        more strings (corresponding to the columns defined by `parse_dates`) as\n",
      "        arguments.\n",
      "    dayfirst : boolean, default False\n",
      "        DD/MM format dates, international and European format\n",
      "    iterator : boolean, default False\n",
      "        Return TextFileReader object for iteration or getting chunks with\n",
      "        ``get_chunk()``.\n",
      "    chunksize : int, default None\n",
      "        Return TextFileReader object for iteration.\n",
      "        See the `IO Tools docs\n",
      "        <http://pandas.pydata.org/pandas-docs/stable/io.html#io-chunking>`_\n",
      "        for more information on ``iterator`` and ``chunksize``.\n",
      "    compression : {'infer', 'gzip', 'bz2', 'zip', 'xz', None}, default 'infer'\n",
      "        For on-the-fly decompression of on-disk data. If 'infer' and\n",
      "        `filepath_or_buffer` is path-like, then detect compression from the\n",
      "        following extensions: '.gz', '.bz2', '.zip', or '.xz' (otherwise no\n",
      "        decompression). If using 'zip', the ZIP file must contain only one data\n",
      "        file to be read in. Set to None for no decompression.\n",
      "    \n",
      "        .. versionadded:: 0.18.1 support for 'zip' and 'xz' compression.\n",
      "    \n",
      "    thousands : str, default None\n",
      "        Thousands separator\n",
      "    decimal : str, default '.'\n",
      "        Character to recognize as decimal point (e.g. use ',' for European data).\n",
      "    float_precision : string, default None\n",
      "        Specifies which converter the C engine should use for floating-point\n",
      "        values. The options are `None` for the ordinary converter,\n",
      "        `high` for the high-precision converter, and `round_trip` for the\n",
      "        round-trip converter.\n",
      "    lineterminator : str (length 1), default None\n",
      "        Character to break file into lines. Only valid with C parser.\n",
      "    quotechar : str (length 1), optional\n",
      "        The character used to denote the start and end of a quoted item. Quoted\n",
      "        items can include the delimiter and it will be ignored.\n",
      "    quoting : int or csv.QUOTE_* instance, default 0\n",
      "        Control field quoting behavior per ``csv.QUOTE_*`` constants. Use one of\n",
      "        QUOTE_MINIMAL (0), QUOTE_ALL (1), QUOTE_NONNUMERIC (2) or QUOTE_NONE (3).\n",
      "    doublequote : boolean, default ``True``\n",
      "       When quotechar is specified and quoting is not ``QUOTE_NONE``, indicate\n",
      "       whether or not to interpret two consecutive quotechar elements INSIDE a\n",
      "       field as a single ``quotechar`` element.\n",
      "    escapechar : str (length 1), default None\n",
      "        One-character string used to escape delimiter when quoting is QUOTE_NONE.\n",
      "    comment : str, default None\n",
      "        Indicates remainder of line should not be parsed. If found at the beginning\n",
      "        of a line, the line will be ignored altogether. This parameter must be a\n",
      "        single character. Like empty lines (as long as ``skip_blank_lines=True``),\n",
      "        fully commented lines are ignored by the parameter `header` but not by\n",
      "        `skiprows`. For example, if comment='#', parsing '#empty\\na,b,c\\n1,2,3'\n",
      "        with `header=0` will result in 'a,b,c' being\n",
      "        treated as the header.\n",
      "    encoding : str, default None\n",
      "        Encoding to use for UTF when reading/writing (ex. 'utf-8'). `List of Python\n",
      "        standard encodings\n",
      "        <https://docs.python.org/3/library/codecs.html#standard-encodings>`_\n",
      "    dialect : str or csv.Dialect instance, default None\n",
      "        If provided, this parameter will override values (default or not) for the\n",
      "        following parameters: `delimiter`, `doublequote`, `escapechar`,\n",
      "        `skipinitialspace`, `quotechar`, and `quoting`. If it is necessary to\n",
      "        override values, a ParserWarning will be issued. See csv.Dialect\n",
      "        documentation for more details.\n",
      "    tupleize_cols : boolean, default False\n",
      "        .. deprecated:: 0.21.0\n",
      "           This argument will be removed and will always convert to MultiIndex\n",
      "    \n",
      "        Leave a list of tuples on columns as is (default is to convert to\n",
      "        a MultiIndex on the columns)\n",
      "    error_bad_lines : boolean, default True\n",
      "        Lines with too many fields (e.g. a csv line with too many commas) will by\n",
      "        default cause an exception to be raised, and no DataFrame will be returned.\n",
      "        If False, then these \"bad lines\" will dropped from the DataFrame that is\n",
      "        returned.\n",
      "    warn_bad_lines : boolean, default True\n",
      "        If error_bad_lines is False, and warn_bad_lines is True, a warning for each\n",
      "        \"bad line\" will be output.\n",
      "    low_memory : boolean, default True\n",
      "        Internally process the file in chunks, resulting in lower memory use\n",
      "        while parsing, but possibly mixed type inference.  To ensure no mixed\n",
      "        types either set False, or specify the type with the `dtype` parameter.\n",
      "        Note that the entire file is read into a single DataFrame regardless,\n",
      "        use the `chunksize` or `iterator` parameter to return the data in chunks.\n",
      "        (Only valid with C parser)\n",
      "    buffer_lines : int, default None\n",
      "        .. deprecated:: 0.19.0\n",
      "           This argument is not respected by the parser\n",
      "    compact_ints : boolean, default False\n",
      "        .. deprecated:: 0.19.0\n",
      "           Argument moved to ``pd.to_numeric``\n",
      "    \n",
      "        If compact_ints is True, then for any column that is of integer dtype,\n",
      "        the parser will attempt to cast it as the smallest integer dtype possible,\n",
      "        either signed or unsigned depending on the specification from the\n",
      "        `use_unsigned` parameter.\n",
      "    use_unsigned : boolean, default False\n",
      "        .. deprecated:: 0.19.0\n",
      "           Argument moved to ``pd.to_numeric``\n",
      "    \n",
      "        If integer columns are being compacted (i.e. `compact_ints=True`), specify\n",
      "        whether the column should be compacted to the smallest signed or unsigned\n",
      "        integer dtype.\n",
      "    memory_map : boolean, default False\n",
      "        If a filepath is provided for `filepath_or_buffer`, map the file object\n",
      "        directly onto memory and access the data directly from there. Using this\n",
      "        option can improve performance because there is no longer any I/O overhead.\n",
      "    \n",
      "    Returns\n",
      "    -------\n",
      "    result : DataFrame or TextParser\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(pd.read_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations_reader = pd.read_table('~/results/all.BRCA-EU.mutations-context.tsv', \n",
    "                                 header=1, \n",
    "                                 delim_whitespace=True,\n",
    "                                 iterator=True)  # Lazily load lines since file may be large"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MUTATION_ID</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>POSITION_GRCh37</th>\n",
       "      <th>POSITION_GRCh38</th>\n",
       "      <th>RELATIVE_POSITION</th>\n",
       "      <th>OVERLAPPED_GENES</th>\n",
       "      <th>CONSEQUENCE(S)</th>\n",
       "      <th>PROJECT(S)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MU64868974</td>\n",
       "      <td>C&gt;G</td>\n",
       "      <td>chr2:169922536</td>\n",
       "      <td>chr2:169066026</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant,ENSG0000...</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MU67221130</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>chr2:1699231</td>\n",
       "      <td>chr2:1695459</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000130508(PXDN)</td>\n",
       "      <td>ENSG00000130508(PXDN):intron_variant,ENSG00000...</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MU64619292</td>\n",
       "      <td>T&gt;-</td>\n",
       "      <td>chr2:16992349</td>\n",
       "      <td>chr2:16811082</td>\n",
       "      <td>INTERGENIC</td>\n",
       "      <td>BRCA-EU</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MU66012277</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>chr2:169923756</td>\n",
       "      <td>chr2:169067246</td>\n",
       "      <td>NON-CODING-EXONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant,ENSG0000...</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MU66538864</td>\n",
       "      <td>A&gt;T</td>\n",
       "      <td>chr2:169925859</td>\n",
       "      <td>chr2:169069349</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant,ENSG0000...</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MU63433685</td>\n",
       "      <td>T&gt;A</td>\n",
       "      <td>chr2:169926262</td>\n",
       "      <td>chr2:169069752</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant,ENSG0000...</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MU64418202</td>\n",
       "      <td>C&gt;G</td>\n",
       "      <td>chr2:169926829</td>\n",
       "      <td>chr2:169070319</td>\n",
       "      <td>NON-CODING-EXONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):5_prime_UTR_variant,ENS...</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MU66019785</td>\n",
       "      <td>T&gt;C</td>\n",
       "      <td>chr2:169927702</td>\n",
       "      <td>chr2:169071192</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant,ENSG0000...</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MU63584668</td>\n",
       "      <td>G&gt;A</td>\n",
       "      <td>chr2:16992866</td>\n",
       "      <td>chr2:16811599</td>\n",
       "      <td>INTERGENIC</td>\n",
       "      <td>BRCA-EU</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MU66013896</td>\n",
       "      <td>T&gt;G</td>\n",
       "      <td>chr2:169928818</td>\n",
       "      <td>chr2:169072308</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant,ENSG0000...</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>MU67196340</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>chr2:169929337</td>\n",
       "      <td>chr2:169072827</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>MU66105214</td>\n",
       "      <td>C&gt;G</td>\n",
       "      <td>chr2:169929410</td>\n",
       "      <td>chr2:169072900</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>MU67026903</td>\n",
       "      <td>T&gt;C</td>\n",
       "      <td>chr2:169931082</td>\n",
       "      <td>chr2:169074572</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>MU65212871</td>\n",
       "      <td>C&gt;G</td>\n",
       "      <td>chr2:169931691</td>\n",
       "      <td>chr2:169075181</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>MU66439690</td>\n",
       "      <td>G&gt;A</td>\n",
       "      <td>chr2:169933547</td>\n",
       "      <td>chr2:169077037</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant,ENSG0000...</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>MU63790286</td>\n",
       "      <td>G&gt;A</td>\n",
       "      <td>chr2:169933977</td>\n",
       "      <td>chr2:169077467</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant,ENSG0000...</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>MU65263602</td>\n",
       "      <td>A&gt;G</td>\n",
       "      <td>chr2:169934557</td>\n",
       "      <td>chr2:169078047</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant,ENSG0000...</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>MU64627748</td>\n",
       "      <td>C&gt;A</td>\n",
       "      <td>chr2:169935171</td>\n",
       "      <td>chr2:169078661</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant,ENSG0000...</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>MU66502389</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>chr2:16993547</td>\n",
       "      <td>chr2:16812280</td>\n",
       "      <td>INTERGENIC</td>\n",
       "      <td>BRCA-EU</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>MU63506559</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>chr2:169936089</td>\n",
       "      <td>chr2:169079579</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant,ENSG0000...</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>MU65823313</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>chr2:1699363</td>\n",
       "      <td>chr2:1695591</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000130508(PXDN)</td>\n",
       "      <td>ENSG00000130508(PXDN):intron_variant,ENSG00000...</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>MU65742206</td>\n",
       "      <td>G&gt;A</td>\n",
       "      <td>chr2:169936965</td>\n",
       "      <td>chr2:169080455</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant,ENSG0000...</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>MU66303273</td>\n",
       "      <td>G&gt;T</td>\n",
       "      <td>chr2:169938689</td>\n",
       "      <td>chr2:169082179</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant,ENSG0000...</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>MU65168306</td>\n",
       "      <td>A&gt;T</td>\n",
       "      <td>chr2:169939183</td>\n",
       "      <td>chr2:169082673</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant,ENSG0000...</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>MU66279019</td>\n",
       "      <td>T&gt;A</td>\n",
       "      <td>chr2:169940971</td>\n",
       "      <td>chr2:169084461</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant,ENSG0000...</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>MU38161712</td>\n",
       "      <td>TC&gt;-</td>\n",
       "      <td>chr2:169942588</td>\n",
       "      <td>chr2:169086078</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant,ENSG0000...</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>MU66372587</td>\n",
       "      <td>T&gt;C</td>\n",
       "      <td>chr2:169942820</td>\n",
       "      <td>chr2:169086310</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant,ENSG0000...</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>MU65451934</td>\n",
       "      <td>T&gt;G</td>\n",
       "      <td>chr2:169944523</td>\n",
       "      <td>chr2:169088013</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>MU64056849</td>\n",
       "      <td>G&gt;T</td>\n",
       "      <td>chr2:169945590</td>\n",
       "      <td>chr2:169089080</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>MU66124010</td>\n",
       "      <td>G&gt;T</td>\n",
       "      <td>chr2:169947354</td>\n",
       "      <td>chr2:169090844</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>MU67026905</td>\n",
       "      <td>G&gt;A</td>\n",
       "      <td>chr2:169947378</td>\n",
       "      <td>chr2:169090868</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>MU63458527</td>\n",
       "      <td>A&gt;T</td>\n",
       "      <td>chr2:169948217</td>\n",
       "      <td>chr2:169091707</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>MU67267855</td>\n",
       "      <td>T&gt;C</td>\n",
       "      <td>chr2:169948700</td>\n",
       "      <td>chr2:169092190</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>MU66460323</td>\n",
       "      <td>G&gt;C</td>\n",
       "      <td>chr2:169948815</td>\n",
       "      <td>chr2:169092305</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>MU63655124</td>\n",
       "      <td>C&gt;A</td>\n",
       "      <td>chr2:169949077</td>\n",
       "      <td>chr2:169092567</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>MU65127267</td>\n",
       "      <td>A&gt;C</td>\n",
       "      <td>chr2:16994913</td>\n",
       "      <td>chr2:16813646</td>\n",
       "      <td>INTERGENIC</td>\n",
       "      <td>BRCA-EU</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>MU66404090</td>\n",
       "      <td>G&gt;C</td>\n",
       "      <td>chr2:169949402</td>\n",
       "      <td>chr2:169092892</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>MU65617357</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>chr2:169949583</td>\n",
       "      <td>chr2:169093073</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>MU2002768</td>\n",
       "      <td>G&gt;C</td>\n",
       "      <td>chr2:169950410</td>\n",
       "      <td>chr2:169093900</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>MU65633152</td>\n",
       "      <td>C&gt;G</td>\n",
       "      <td>chr2:169951193</td>\n",
       "      <td>chr2:169094683</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>MU64997961</td>\n",
       "      <td>G&gt;A</td>\n",
       "      <td>chr2:169951824</td>\n",
       "      <td>chr2:169095314</td>\n",
       "      <td>INTRONIC</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):intron_variant</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>MU65445163</td>\n",
       "      <td>C&gt;G</td>\n",
       "      <td>chr2:169952242</td>\n",
       "      <td>chr2:169095732</td>\n",
       "      <td>EXONIC:0</td>\n",
       "      <td>ENSG00000073737(DHRS9)</td>\n",
       "      <td>ENSG00000073737(DHRS9):missense_variant</td>\n",
       "      <td>BRCA-EU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>MU66456945</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>chr2:169954085</td>\n",
       "      <td>chr2:169097575</td>\n",
       "      <td>INTERGENIC</td>\n",
       "      <td>ENSG00000073737(DHRS9):downstream_gene_variant...</td>\n",
       "      <td>BRCA-EU</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>MU66439692</td>\n",
       "      <td>T&gt;C</td>\n",
       "      <td>chr2:169954735</td>\n",
       "      <td>chr2:169098225</td>\n",
       "      <td>INTERGENIC</td>\n",
       "      <td>ENSG00000073737(DHRS9):downstream_gene_variant...</td>\n",
       "      <td>BRCA-EU</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>MU66595122</td>\n",
       "      <td>G&gt;A</td>\n",
       "      <td>chr2:169955219</td>\n",
       "      <td>chr2:169098709</td>\n",
       "      <td>INTERGENIC</td>\n",
       "      <td>ENSG00000073737(DHRS9):downstream_gene_variant...</td>\n",
       "      <td>BRCA-EU</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>MU65909817</td>\n",
       "      <td>C&gt;A</td>\n",
       "      <td>chr2:169955404</td>\n",
       "      <td>chr2:169098894</td>\n",
       "      <td>INTERGENIC</td>\n",
       "      <td>ENSG00000073737(DHRS9):downstream_gene_variant...</td>\n",
       "      <td>BRCA-EU</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>MU64961048</td>\n",
       "      <td>A&gt;T</td>\n",
       "      <td>chr2:169955614</td>\n",
       "      <td>chr2:169099104</td>\n",
       "      <td>INTERGENIC</td>\n",
       "      <td>ENSG00000073737(DHRS9):downstream_gene_variant...</td>\n",
       "      <td>BRCA-EU</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>MU64997969</td>\n",
       "      <td>G&gt;A</td>\n",
       "      <td>chr2:169955793</td>\n",
       "      <td>chr2:169099283</td>\n",
       "      <td>INTERGENIC</td>\n",
       "      <td>ENSG00000073737(DHRS9):downstream_gene_variant...</td>\n",
       "      <td>BRCA-EU</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>MU63655130</td>\n",
       "      <td>A&gt;T</td>\n",
       "      <td>chr2:169956611</td>\n",
       "      <td>chr2:169100101</td>\n",
       "      <td>INTERGENIC</td>\n",
       "      <td>ENSG00000073737(DHRS9):downstream_gene_variant...</td>\n",
       "      <td>BRCA-EU</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>MU65400707</td>\n",
       "      <td>G&gt;C</td>\n",
       "      <td>chr2:169957921</td>\n",
       "      <td>chr2:169101411</td>\n",
       "      <td>INTERGENIC</td>\n",
       "      <td>ENSG00000235321(AC007556.3):downstream_gene_va...</td>\n",
       "      <td>BRCA-EU</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   MUTATION_ID MUTATION POSITION_GRCh37 POSITION_GRCh38  RELATIVE_POSITION  \\\n",
       "0   MU64868974      C>G  chr2:169922536  chr2:169066026           INTRONIC   \n",
       "1   MU67221130      C>T    chr2:1699231    chr2:1695459           INTRONIC   \n",
       "2   MU64619292      T>-   chr2:16992349   chr2:16811082         INTERGENIC   \n",
       "3   MU66012277      C>T  chr2:169923756  chr2:169067246  NON-CODING-EXONIC   \n",
       "4   MU66538864      A>T  chr2:169925859  chr2:169069349           INTRONIC   \n",
       "5   MU63433685      T>A  chr2:169926262  chr2:169069752           INTRONIC   \n",
       "6   MU64418202      C>G  chr2:169926829  chr2:169070319  NON-CODING-EXONIC   \n",
       "7   MU66019785      T>C  chr2:169927702  chr2:169071192           INTRONIC   \n",
       "8   MU63584668      G>A   chr2:16992866   chr2:16811599         INTERGENIC   \n",
       "9   MU66013896      T>G  chr2:169928818  chr2:169072308           INTRONIC   \n",
       "10  MU67196340      C>T  chr2:169929337  chr2:169072827           INTRONIC   \n",
       "11  MU66105214      C>G  chr2:169929410  chr2:169072900           INTRONIC   \n",
       "12  MU67026903      T>C  chr2:169931082  chr2:169074572           INTRONIC   \n",
       "13  MU65212871      C>G  chr2:169931691  chr2:169075181           INTRONIC   \n",
       "14  MU66439690      G>A  chr2:169933547  chr2:169077037           INTRONIC   \n",
       "15  MU63790286      G>A  chr2:169933977  chr2:169077467           INTRONIC   \n",
       "16  MU65263602      A>G  chr2:169934557  chr2:169078047           INTRONIC   \n",
       "17  MU64627748      C>A  chr2:169935171  chr2:169078661           INTRONIC   \n",
       "18  MU66502389      C>T   chr2:16993547   chr2:16812280         INTERGENIC   \n",
       "19  MU63506559      C>T  chr2:169936089  chr2:169079579           INTRONIC   \n",
       "20  MU65823313      C>T    chr2:1699363    chr2:1695591           INTRONIC   \n",
       "21  MU65742206      G>A  chr2:169936965  chr2:169080455           INTRONIC   \n",
       "22  MU66303273      G>T  chr2:169938689  chr2:169082179           INTRONIC   \n",
       "23  MU65168306      A>T  chr2:169939183  chr2:169082673           INTRONIC   \n",
       "24  MU66279019      T>A  chr2:169940971  chr2:169084461           INTRONIC   \n",
       "25  MU38161712     TC>-  chr2:169942588  chr2:169086078           INTRONIC   \n",
       "26  MU66372587      T>C  chr2:169942820  chr2:169086310           INTRONIC   \n",
       "27  MU65451934      T>G  chr2:169944523  chr2:169088013           INTRONIC   \n",
       "28  MU64056849      G>T  chr2:169945590  chr2:169089080           INTRONIC   \n",
       "29  MU66124010      G>T  chr2:169947354  chr2:169090844           INTRONIC   \n",
       "30  MU67026905      G>A  chr2:169947378  chr2:169090868           INTRONIC   \n",
       "31  MU63458527      A>T  chr2:169948217  chr2:169091707           INTRONIC   \n",
       "32  MU67267855      T>C  chr2:169948700  chr2:169092190           INTRONIC   \n",
       "33  MU66460323      G>C  chr2:169948815  chr2:169092305           INTRONIC   \n",
       "34  MU63655124      C>A  chr2:169949077  chr2:169092567           INTRONIC   \n",
       "35  MU65127267      A>C   chr2:16994913   chr2:16813646         INTERGENIC   \n",
       "36  MU66404090      G>C  chr2:169949402  chr2:169092892           INTRONIC   \n",
       "37  MU65617357      C>T  chr2:169949583  chr2:169093073           INTRONIC   \n",
       "38   MU2002768      G>C  chr2:169950410  chr2:169093900           INTRONIC   \n",
       "39  MU65633152      C>G  chr2:169951193  chr2:169094683           INTRONIC   \n",
       "40  MU64997961      G>A  chr2:169951824  chr2:169095314           INTRONIC   \n",
       "41  MU65445163      C>G  chr2:169952242  chr2:169095732           EXONIC:0   \n",
       "42  MU66456945      C>T  chr2:169954085  chr2:169097575         INTERGENIC   \n",
       "43  MU66439692      T>C  chr2:169954735  chr2:169098225         INTERGENIC   \n",
       "44  MU66595122      G>A  chr2:169955219  chr2:169098709         INTERGENIC   \n",
       "45  MU65909817      C>A  chr2:169955404  chr2:169098894         INTERGENIC   \n",
       "46  MU64961048      A>T  chr2:169955614  chr2:169099104         INTERGENIC   \n",
       "47  MU64997969      G>A  chr2:169955793  chr2:169099283         INTERGENIC   \n",
       "48  MU63655130      A>T  chr2:169956611  chr2:169100101         INTERGENIC   \n",
       "49  MU65400707      G>C  chr2:169957921  chr2:169101411         INTERGENIC   \n",
       "\n",
       "                                     OVERLAPPED_GENES  \\\n",
       "0                              ENSG00000073737(DHRS9)   \n",
       "1                               ENSG00000130508(PXDN)   \n",
       "2                                             BRCA-EU   \n",
       "3                              ENSG00000073737(DHRS9)   \n",
       "4                              ENSG00000073737(DHRS9)   \n",
       "5                              ENSG00000073737(DHRS9)   \n",
       "6                              ENSG00000073737(DHRS9)   \n",
       "7                              ENSG00000073737(DHRS9)   \n",
       "8                                             BRCA-EU   \n",
       "9                              ENSG00000073737(DHRS9)   \n",
       "10                             ENSG00000073737(DHRS9)   \n",
       "11                             ENSG00000073737(DHRS9)   \n",
       "12                             ENSG00000073737(DHRS9)   \n",
       "13                             ENSG00000073737(DHRS9)   \n",
       "14                             ENSG00000073737(DHRS9)   \n",
       "15                             ENSG00000073737(DHRS9)   \n",
       "16                             ENSG00000073737(DHRS9)   \n",
       "17                             ENSG00000073737(DHRS9)   \n",
       "18                                            BRCA-EU   \n",
       "19                             ENSG00000073737(DHRS9)   \n",
       "20                              ENSG00000130508(PXDN)   \n",
       "21                             ENSG00000073737(DHRS9)   \n",
       "22                             ENSG00000073737(DHRS9)   \n",
       "23                             ENSG00000073737(DHRS9)   \n",
       "24                             ENSG00000073737(DHRS9)   \n",
       "25                             ENSG00000073737(DHRS9)   \n",
       "26                             ENSG00000073737(DHRS9)   \n",
       "27                             ENSG00000073737(DHRS9)   \n",
       "28                             ENSG00000073737(DHRS9)   \n",
       "29                             ENSG00000073737(DHRS9)   \n",
       "30                             ENSG00000073737(DHRS9)   \n",
       "31                             ENSG00000073737(DHRS9)   \n",
       "32                             ENSG00000073737(DHRS9)   \n",
       "33                             ENSG00000073737(DHRS9)   \n",
       "34                             ENSG00000073737(DHRS9)   \n",
       "35                                            BRCA-EU   \n",
       "36                             ENSG00000073737(DHRS9)   \n",
       "37                             ENSG00000073737(DHRS9)   \n",
       "38                             ENSG00000073737(DHRS9)   \n",
       "39                             ENSG00000073737(DHRS9)   \n",
       "40                             ENSG00000073737(DHRS9)   \n",
       "41                             ENSG00000073737(DHRS9)   \n",
       "42  ENSG00000073737(DHRS9):downstream_gene_variant...   \n",
       "43  ENSG00000073737(DHRS9):downstream_gene_variant...   \n",
       "44  ENSG00000073737(DHRS9):downstream_gene_variant...   \n",
       "45  ENSG00000073737(DHRS9):downstream_gene_variant...   \n",
       "46  ENSG00000073737(DHRS9):downstream_gene_variant...   \n",
       "47  ENSG00000073737(DHRS9):downstream_gene_variant...   \n",
       "48  ENSG00000073737(DHRS9):downstream_gene_variant...   \n",
       "49  ENSG00000235321(AC007556.3):downstream_gene_va...   \n",
       "\n",
       "                                       CONSEQUENCE(S) PROJECT(S)  \n",
       "0   ENSG00000073737(DHRS9):intron_variant,ENSG0000...    BRCA-EU  \n",
       "1   ENSG00000130508(PXDN):intron_variant,ENSG00000...    BRCA-EU  \n",
       "2                                                 NaN        NaN  \n",
       "3   ENSG00000073737(DHRS9):intron_variant,ENSG0000...    BRCA-EU  \n",
       "4   ENSG00000073737(DHRS9):intron_variant,ENSG0000...    BRCA-EU  \n",
       "5   ENSG00000073737(DHRS9):intron_variant,ENSG0000...    BRCA-EU  \n",
       "6   ENSG00000073737(DHRS9):5_prime_UTR_variant,ENS...    BRCA-EU  \n",
       "7   ENSG00000073737(DHRS9):intron_variant,ENSG0000...    BRCA-EU  \n",
       "8                                                 NaN        NaN  \n",
       "9   ENSG00000073737(DHRS9):intron_variant,ENSG0000...    BRCA-EU  \n",
       "10              ENSG00000073737(DHRS9):intron_variant    BRCA-EU  \n",
       "11              ENSG00000073737(DHRS9):intron_variant    BRCA-EU  \n",
       "12              ENSG00000073737(DHRS9):intron_variant    BRCA-EU  \n",
       "13              ENSG00000073737(DHRS9):intron_variant    BRCA-EU  \n",
       "14  ENSG00000073737(DHRS9):intron_variant,ENSG0000...    BRCA-EU  \n",
       "15  ENSG00000073737(DHRS9):intron_variant,ENSG0000...    BRCA-EU  \n",
       "16  ENSG00000073737(DHRS9):intron_variant,ENSG0000...    BRCA-EU  \n",
       "17  ENSG00000073737(DHRS9):intron_variant,ENSG0000...    BRCA-EU  \n",
       "18                                                NaN        NaN  \n",
       "19  ENSG00000073737(DHRS9):intron_variant,ENSG0000...    BRCA-EU  \n",
       "20  ENSG00000130508(PXDN):intron_variant,ENSG00000...    BRCA-EU  \n",
       "21  ENSG00000073737(DHRS9):intron_variant,ENSG0000...    BRCA-EU  \n",
       "22  ENSG00000073737(DHRS9):intron_variant,ENSG0000...    BRCA-EU  \n",
       "23  ENSG00000073737(DHRS9):intron_variant,ENSG0000...    BRCA-EU  \n",
       "24  ENSG00000073737(DHRS9):intron_variant,ENSG0000...    BRCA-EU  \n",
       "25  ENSG00000073737(DHRS9):intron_variant,ENSG0000...    BRCA-EU  \n",
       "26  ENSG00000073737(DHRS9):intron_variant,ENSG0000...    BRCA-EU  \n",
       "27              ENSG00000073737(DHRS9):intron_variant    BRCA-EU  \n",
       "28              ENSG00000073737(DHRS9):intron_variant    BRCA-EU  \n",
       "29              ENSG00000073737(DHRS9):intron_variant    BRCA-EU  \n",
       "30              ENSG00000073737(DHRS9):intron_variant    BRCA-EU  \n",
       "31              ENSG00000073737(DHRS9):intron_variant    BRCA-EU  \n",
       "32              ENSG00000073737(DHRS9):intron_variant    BRCA-EU  \n",
       "33              ENSG00000073737(DHRS9):intron_variant    BRCA-EU  \n",
       "34              ENSG00000073737(DHRS9):intron_variant    BRCA-EU  \n",
       "35                                                NaN        NaN  \n",
       "36              ENSG00000073737(DHRS9):intron_variant    BRCA-EU  \n",
       "37              ENSG00000073737(DHRS9):intron_variant    BRCA-EU  \n",
       "38              ENSG00000073737(DHRS9):intron_variant    BRCA-EU  \n",
       "39              ENSG00000073737(DHRS9):intron_variant    BRCA-EU  \n",
       "40              ENSG00000073737(DHRS9):intron_variant    BRCA-EU  \n",
       "41            ENSG00000073737(DHRS9):missense_variant    BRCA-EU  \n",
       "42                                            BRCA-EU        NaN  \n",
       "43                                            BRCA-EU        NaN  \n",
       "44                                            BRCA-EU        NaN  \n",
       "45                                            BRCA-EU        NaN  \n",
       "46                                            BRCA-EU        NaN  \n",
       "47                                            BRCA-EU        NaN  \n",
       "48                                            BRCA-EU        NaN  \n",
       "49                                            BRCA-EU        NaN  "
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations = mutations_reader.get_chunk(50)\n",
    "mutations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Data mangling\n",
    "============="
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0, 1, 2, 3, 4]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Helper function to test\n",
    "from itertools import islice\n",
    "\n",
    "def head(iterable, items=10):\n",
    "    'Return the first items of an iterator.'\n",
    "    iterator = iter(iterable)\n",
    "    return islice(iterable, items)\n",
    "# ---\n",
    "\n",
    "[item for item in head(range(100), items=5)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pandas(Index=0, MUTATION_ID='MU64868974', MUTATION='C>G', POSITION_GRCh37='chr2:169922536', POSITION_GRCh38='chr2:169066026', RELATIVE_POSITION='INTRONIC', OVERLAPPED_GENES='ENSG00000073737(DHRS9)', _7='ENSG00000073737(DHRS9):intron_variant,ENSG00000073737(DHRS9):upstream_gene_variant', _8='BRCA-EU')\n",
      "chr2:1699231\n",
      "MU64619292\n",
      "MU66012277\n",
      "MU66538864\n",
      "MU63433685\n",
      "MU64418202\n",
      "MU66019785\n",
      "MU63584668\n",
      "MU66013896\n",
      "MU67196340\n",
      "MU66105214\n"
     ]
    }
   ],
   "source": [
    "## Exploring how to iterate over the data\n",
    "i = mutations.itertuples()\n",
    "\n",
    "print(next(i))\n",
    "\n",
    "row = next(i)\n",
    "print(row.POSITION_GRCh37)\n",
    "\n",
    "for mutation in head(i):\n",
    "    print(mutation.MUTATION_ID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MU64868974 2 169922536\n",
      "MU67221130 2 1699231\n",
      "MU64619292 2 16992349\n",
      "MU66012277 2 169923756\n",
      "MU66538864 2 169925859\n",
      "MU63433685 2 169926262\n",
      "MU64418202 2 169926829\n",
      "MU66019785 2 169927702\n",
      "MU63584668 2 16992866\n",
      "MU66013896 2 169928818\n"
     ]
    }
   ],
   "source": [
    "# Test on the file\n",
    "for mutation in head(mutations.itertuples()):\n",
    "    \n",
    "    # < --- Parse the position \n",
    "    chrom_str, pos_str = mutation.POSITION_GRCh37.split(':')\n",
    "    # For the chromosome, eliminate the trailing \"chr\"\n",
    "    chrom = chrom_str.replace('chr', '')\n",
    "    # The position is an integer\n",
    "    pos = int(pos_str)\n",
    "    \n",
    "    print(mutation.MUTATION_ID, chrom, pos)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Exploring how to validate\n",
    "===============\n",
    "\n",
    "The validation is done by requests to the [Ensembl REST API](http://rest.ensembl.org/documentation/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests, sys\n",
    " \n",
    "grch38_server = \"http://rest.ensembl.org\"\n",
    "server = \"http://grch37.rest.ensembl.org\" # We are using the GRCh37 server\n",
    "\n",
    "\n",
    "ext = \"/sequence/region/human\"\n",
    "headers={ \"Content-Type\" : \"application/json\", \n",
    "         \"Accept\" : \"application/json\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'query': 'X:1000000..1000100:1', 'id': 'chromosome:GRCh37:X:1000000:1000100:1', 'seq': 'GAAACAGCTACTTGGAAGGCTGAAGCAGGAGGATTGTTTGAGTCTAGGAGTTTGAGGCTGCAGTGAGTTATGAGCACACCACGGCACTCCAGCCTGGGAGA', 'molecule': 'dna'}, {'query': 'ABBA01004489.1:1..100', 'id': 'contig::ABBA01004489.1:1:100:1', 'seq': 'CTGTACTTTCCTTGGGATGGAGTAGTTTCGAAACACACTTTCTGTAGAATCTGCAAGTGGATATTTGGACCTGTCTGAGGAATTCGTTGGAAACGGGATA', 'molecule': 'dna'}]\n"
     ]
    }
   ],
   "source": [
    "r = requests.post(server+ext, \n",
    "                  headers=headers, \n",
    "                  data='''{\"regions\":[\"X:1000000..1000100:1\", \n",
    "                                    \"ABBA01004489.1:1..100\"] }''')\n",
    " \n",
    "if not r.ok:\n",
    "  r.raise_for_status()\n",
    "  sys.exit()\n",
    " \n",
    "decoded = r.json()\n",
    "print(repr(decoded))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2:169922536..169922536\n",
      "2:1699231..1699231\n",
      "2:16992349..16992349\n",
      "2:169923756..169923756\n",
      "2:169925859..169925859\n",
      "2:169926262..169926262\n",
      "2:169926829..169926829\n",
      "2:169927702..169927702\n",
      "2:16992866..16992866\n",
      "2:169928818..169928818\n"
     ]
    }
   ],
   "source": [
    "#### Test for region string encoding\n",
    "\n",
    "for mutation in head(mutations.itertuples()):\n",
    "    \n",
    "    # < --- Parse the mutation's start position \n",
    "    chrom_str, pos_str = mutation.POSITION_GRCh37.split(':')\n",
    "    # For the chromosome, eliminate the trailing \"chr\"\n",
    "    chrom_str = chrom_str.replace('chr', '')\n",
    "    # The position is an integer\n",
    "    pos = int(pos_str)\n",
    "    \n",
    "    # < --- Parse the mutation's end position\n",
    "    # Get the mutation's reference allele\n",
    "    ref, _ = mutation.MUTATION.split('>')\n",
    "    end = pos + (len(ref)-1)\n",
    "    \n",
    "    # < --- Assemble the region string\n",
    "    region_str = '{chrom}:{start}..{end}'.format(chrom=chrom_str, \n",
    "                                                 start=pos_str, \n",
    "                                                 end=end)\n",
    "    print(region_str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2:169922535..169922537\n",
      "2:1699230..1699232\n",
      "2:16992348..16992350\n",
      "2:169923755..169923757\n",
      "2:169925858..169925860\n",
      "2:169926261..169926263\n",
      "2:169926828..169926830\n",
      "2:169927701..169927703\n",
      "2:16992865..16992867\n",
      "2:169928817..169928819\n"
     ]
    }
   ],
   "source": [
    "####  Assemble that into a function\n",
    "\n",
    "def to_region_str(mutation):\n",
    "    'Get a region string suitable for requests to Ensembl'\n",
    "    \n",
    "    # < --- Parse the mutation's start position \n",
    "    chrom_str, pos_str = mutation.POSITION_GRCh37.split(':')\n",
    "    # For the chromosome, eliminate the trailing \"chr\"\n",
    "    chrom_str = chrom_str.replace('chr', '')\n",
    "    # The position is an integer\n",
    "    pos = int(pos_str)\n",
    "    \n",
    "    # < --- Parse the mutation's end position\n",
    "    # Get the mutation's reference allele\n",
    "    ref, _ = mutation.MUTATION.split('>')\n",
    "    end = pos + (len(ref)-1)\n",
    "    \n",
    "    # < --- Assemble the region string\n",
    "    return '{chrom}:{start}..{end}'.format(chrom=chrom_str, \n",
    "                                           start=pos-1, # Just in case, extra space\n",
    "                                           end=end+1) # Also, in case\n",
    "# ---\n",
    "\n",
    "for mutation in head(mutations.itertuples()):\n",
    "    print(to_region_str(mutation))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'{\"regions\": [\"2:169922535..169922537\", \"2:1699230..1699232\", \"2:16992348..16992350\", \"2:169923755..169923757\", \"2:169925858..169925860\", \"2:169926261..169926263\", \"2:169926828..169926830\", \"2:169927701..169927703\", \"2:16992865..16992867\", \"2:169928817..169928819\"]}'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "####  Test for JSON string encoding\n",
    "\n",
    "import json\n",
    "from collections import defaultdict\n",
    "\n",
    "data = defaultdict(list)\n",
    "for mutation in head(mutations.itertuples()):\n",
    "    # Add region string to the data dictionary\n",
    "    # at the automatically created 'region' key\n",
    "    data['regions'].append( to_region_str(mutation) )\n",
    "\n",
    "json.dumps(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'id': 'chromosome:GRCh37:2:169922535:169922537:1',\n",
       "  'molecule': 'dna',\n",
       "  'query': '2:169922535..169922537',\n",
       "  'seq': 'TCC'},\n",
       " {'id': 'chromosome:GRCh37:2:1699230:1699232:1',\n",
       "  'molecule': 'dna',\n",
       "  'query': '2:1699230..1699232',\n",
       "  'seq': 'TCC'},\n",
       " {'id': 'chromosome:GRCh37:2:16992348:16992350:1',\n",
       "  'molecule': 'dna',\n",
       "  'query': '2:16992348..16992350',\n",
       "  'seq': 'TAT'},\n",
       " {'id': 'chromosome:GRCh37:2:169923755:169923757:1',\n",
       "  'molecule': 'dna',\n",
       "  'query': '2:169923755..169923757',\n",
       "  'seq': 'CCC'},\n",
       " {'id': 'chromosome:GRCh37:2:169925858:169925860:1',\n",
       "  'molecule': 'dna',\n",
       "  'query': '2:169925858..169925860',\n",
       "  'seq': 'GAG'},\n",
       " {'id': 'chromosome:GRCh37:2:169926261:169926263:1',\n",
       "  'molecule': 'dna',\n",
       "  'query': '2:169926261..169926263',\n",
       "  'seq': 'ATT'},\n",
       " {'id': 'chromosome:GRCh37:2:169926828:169926830:1',\n",
       "  'molecule': 'dna',\n",
       "  'query': '2:169926828..169926830',\n",
       "  'seq': 'ACT'},\n",
       " {'id': 'chromosome:GRCh37:2:169927701:169927703:1',\n",
       "  'molecule': 'dna',\n",
       "  'query': '2:169927701..169927703',\n",
       "  'seq': 'TTC'},\n",
       " {'id': 'chromosome:GRCh37:2:16992865:16992867:1',\n",
       "  'molecule': 'dna',\n",
       "  'query': '2:16992865..16992867',\n",
       "  'seq': 'TGT'},\n",
       " {'id': 'chromosome:GRCh37:2:169928817:169928819:1',\n",
       "  'molecule': 'dna',\n",
       "  'query': '2:169928817..169928819',\n",
       "  'seq': 'TTC'}]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#### First actual requests\n",
    "\n",
    "import json\n",
    "from collections import defaultdict\n",
    "\n",
    "\n",
    "# < --- Assemble the request data\n",
    "\n",
    "request_data = defaultdict(list)\n",
    "for mutation in head(mutations.itertuples()):\n",
    "    # Add region string to the data dictionary\n",
    "    # at the automatically created 'region' key\n",
    "    request_data['regions'].append( to_region_str(mutation) )\n",
    "\n",
    "# < --- Make the request\n",
    "\n",
    "# Request the sequences at the position\n",
    "r = requests.post(server+ext, \n",
    "                  headers=headers, \n",
    "                  data=json.dumps(data)) # Here goes the data\n",
    "\n",
    "# < --- Check the response\n",
    "\n",
    "if not r.ok:\n",
    "    r.raise_for_status()\n",
    " \n",
    "r.json()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TCC\n",
      "TCC\n",
      "TAT\n",
      "CCC\n",
      "GAG\n",
      "ATT\n",
      "ACT\n",
      "TTC\n",
      "TGT\n",
      "TTC\n"
     ]
    }
   ],
   "source": [
    "#### Inspect the response\n",
    "for item in r.json():\n",
    "    print(item['seq'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TCC C\n",
      "TCC C\n",
      "TAT T\n",
      "CCC C\n",
      "GAG A\n",
      "ATT T\n",
      "ACT C\n",
      "TTC T\n",
      "TGT G\n",
      "TTC T\n"
     ]
    }
   ],
   "source": [
    "#### Verify first positions\n",
    "for r_item, mut in zip(r.json(), mutations.itertuples()):\n",
    "    print(r_item['seq'], mut.MUTATION.split('>')[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Yay! It worked! Now let's automate the validation process."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, the `to_region_str` fetched extra positions on purpose. So, let's remove that. Besides, it wolu be good to be able to specify whether we are validating GRCh37 or GRCh38 positions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2:169922535..169922537\n",
      "2:1699230..1699232\n",
      "2:16992348..16992350\n",
      "2:169923755..169923757\n",
      "2:169925858..169925860\n",
      "2:169926261..169926263\n",
      "2:169926828..169926830\n",
      "2:169927701..169927703\n",
      "2:16992865..16992867\n",
      "2:169928817..169928819\n"
     ]
    }
   ],
   "source": [
    "def to_region_str(mutation, assembly='GRCh37'):\n",
    "    'Get a region string suitable for requests to Ensembl'\n",
    "    \n",
    "    # < --- Parse the mutation's start position\n",
    "    if assembly == 'GRCh37':\n",
    "        chrom_str, pos_str = mutation.POSITION_GRCh37.split(':')\n",
    "    else:\n",
    "        chrom_str, pos_str = mutation.POSITION_GRCh38.split(':')\n",
    "    # For the chromosome, eliminate the trailing \"chr\"\n",
    "    chrom_str = chrom_str.replace('chr', '')\n",
    "    # The position is an integer\n",
    "    pos = int(pos_str)\n",
    "    \n",
    "    # < --- Parse the mutation's end position\n",
    "    # Get the mutation's reference allele\n",
    "    ref, _ = mutation.MUTATION.split('>')\n",
    "    end = pos + (len(ref)-1)\n",
    "    \n",
    "    # < --- Assemble the region string\n",
    "    return '{chrom}:{start}..{end}'.format(chrom=chrom_str, \n",
    "                                           start=(pos-1), \n",
    "                                           end=(end+1)) # Just the right chars\n",
    "# ---\n",
    "\n",
    "for mutation in head(mutations.itertuples()):\n",
    "    print(to_region_str(mutation))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, let's define the complete algorithm:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID:  MU64868974 Expected:  C Got:  TCC\n",
      "ID:  MU67221130 Expected:  C Got:  TCC\n",
      "ID:  MU64619292 Expected:  T Got:  TAT\n",
      "ID:  MU66012277 Expected:  C Got:  CCC\n",
      "ID:  MU66538864 Expected:  A Got:  GAG\n",
      "ID:  MU63433685 Expected:  T Got:  ATT\n",
      "ID:  MU64418202 Expected:  C Got:  ACT\n",
      "ID:  MU66019785 Expected:  T Got:  TTC\n",
      "ID:  MU63584668 Expected:  G Got:  TGT\n",
      "ID:  MU66013896 Expected:  T Got:  TTC\n",
      "ID:  MU67196340 Expected:  C Got:  TCA\n",
      "ID:  MU66105214 Expected:  C Got:  TCT\n",
      "ID:  MU67026903 Expected:  T Got:  GTT\n",
      "ID:  MU65212871 Expected:  C Got:  TCT\n",
      "ID:  MU66439690 Expected:  G Got:  AGG\n",
      "ID:  MU63790286 Expected:  G Got:  AGA\n",
      "ID:  MU65263602 Expected:  A Got:  CAG\n",
      "ID:  MU64627748 Expected:  C Got:  ACA\n",
      "ID:  MU66502389 Expected:  C Got:  TCA\n",
      "ID:  MU63506559 Expected:  C Got:  GCG\n",
      "ID:  MU65823313 Expected:  C Got:  TCC\n",
      "ID:  MU65742206 Expected:  G Got:  AGG\n",
      "ID:  MU66303273 Expected:  G Got:  GGC\n",
      "ID:  MU65168306 Expected:  A Got:  AAG\n",
      "ID:  MU66279019 Expected:  T Got:  GTG\n",
      "ID:  MU38161712 Expected:  TC Got:  CATC\n",
      "ID:  MU66372587 Expected:  T Got:  ATC\n",
      "ID:  MU65451934 Expected:  T Got:  TTG\n",
      "ID:  MU64056849 Expected:  G Got:  TGG\n",
      "ID:  MU66124010 Expected:  G Got:  AGC\n",
      "ID:  MU67026905 Expected:  G Got:  CGA\n",
      "ID:  MU63458527 Expected:  A Got:  GAT\n",
      "ID:  MU67267855 Expected:  T Got:  TTA\n",
      "ID:  MU66460323 Expected:  G Got:  CGA\n",
      "ID:  MU63655124 Expected:  C Got:  TCC\n",
      "ID:  MU65127267 Expected:  A Got:  AAG\n",
      "ID:  MU66404090 Expected:  G Got:  AGA\n",
      "ID:  MU65617357 Expected:  C Got:  TCT\n",
      "ID:  MU2002768 Expected:  G Got:  AGA\n",
      "ID:  MU65633152 Expected:  C Got:  TCT\n",
      "ID:  MU64997961 Expected:  G Got:  TGA\n",
      "ID:  MU65445163 Expected:  C Got:  ACA\n",
      "ID:  MU66456945 Expected:  C Got:  TCG\n",
      "ID:  MU66439692 Expected:  T Got:  CTA\n",
      "ID:  MU66595122 Expected:  G Got:  GGA\n",
      "ID:  MU65909817 Expected:  C Got:  TCT\n",
      "ID:  MU64961048 Expected:  A Got:  GAA\n",
      "ID:  MU64997969 Expected:  G Got:  TGA\n",
      "ID:  MU63655130 Expected:  A Got:  AAT\n",
      "ID:  MU65400707 Expected:  G Got:  TGA\n",
      "[{'ID': 'MU64619292', 'expected': 'T', 'got': 'TAT', 'pos': 16992349},\n",
      " {'ID': 'MU38161712', 'expected': 'TC', 'got': 'CATC', 'pos': 169942588}]\n",
      "ID:  MU64868974 Expected:  C Got:  TCC\n",
      "ID:  MU67221130 Expected:  C Got:  TCC\n",
      "ID:  MU64619292 Expected:  T Got:  TAT\n",
      "ID:  MU66012277 Expected:  C Got:  CCC\n",
      "ID:  MU66538864 Expected:  A Got:  GAG\n",
      "ID:  MU63433685 Expected:  T Got:  ATT\n",
      "ID:  MU64418202 Expected:  C Got:  ACT\n",
      "ID:  MU66019785 Expected:  T Got:  TTC\n",
      "ID:  MU63584668 Expected:  G Got:  TGT\n",
      "ID:  MU66013896 Expected:  T Got:  TTC\n",
      "ID:  MU67196340 Expected:  C Got:  TCA\n",
      "ID:  MU66105214 Expected:  C Got:  TCT\n",
      "ID:  MU67026903 Expected:  T Got:  GTT\n",
      "ID:  MU65212871 Expected:  C Got:  TCT\n",
      "ID:  MU66439690 Expected:  G Got:  AGG\n",
      "ID:  MU63790286 Expected:  G Got:  AGA\n",
      "ID:  MU65263602 Expected:  A Got:  CAG\n",
      "ID:  MU64627748 Expected:  C Got:  ACA\n",
      "ID:  MU66502389 Expected:  C Got:  TCA\n",
      "ID:  MU63506559 Expected:  C Got:  GCG\n",
      "ID:  MU65823313 Expected:  C Got:  TCC\n",
      "ID:  MU65742206 Expected:  G Got:  AGG\n",
      "ID:  MU66303273 Expected:  G Got:  GGC\n",
      "ID:  MU65168306 Expected:  A Got:  AAG\n",
      "ID:  MU66279019 Expected:  T Got:  GTG\n",
      "ID:  MU38161712 Expected:  TC Got:  CATC\n",
      "ID:  MU66372587 Expected:  T Got:  ATC\n",
      "ID:  MU65451934 Expected:  T Got:  TTG\n",
      "ID:  MU64056849 Expected:  G Got:  TGG\n",
      "ID:  MU66124010 Expected:  G Got:  AGC\n",
      "ID:  MU67026905 Expected:  G Got:  CGA\n",
      "ID:  MU63458527 Expected:  A Got:  GAT\n",
      "ID:  MU67267855 Expected:  T Got:  TTA\n",
      "ID:  MU66460323 Expected:  G Got:  CGA\n",
      "ID:  MU63655124 Expected:  C Got:  TCC\n",
      "ID:  MU65127267 Expected:  A Got:  AAG\n",
      "ID:  MU66404090 Expected:  G Got:  AGA\n",
      "ID:  MU65617357 Expected:  C Got:  TCT\n",
      "ID:  MU2002768 Expected:  G Got:  AGA\n",
      "ID:  MU65633152 Expected:  C Got:  TCT\n",
      "ID:  MU64997961 Expected:  G Got:  TGA\n",
      "ID:  MU65445163 Expected:  C Got:  ACA\n",
      "ID:  MU66456945 Expected:  C Got:  TCG\n",
      "ID:  MU66439692 Expected:  T Got:  CTA\n",
      "ID:  MU66595122 Expected:  G Got:  GGA\n",
      "ID:  MU65909817 Expected:  C Got:  TCT\n",
      "ID:  MU64961048 Expected:  A Got:  GAA\n",
      "ID:  MU64997969 Expected:  G Got:  TGA\n",
      "ID:  MU63655130 Expected:  A Got:  AAT\n",
      "ID:  MU65400707 Expected:  G Got:  TGA\n",
      "[{'ID': 'MU64619292', 'expected': 'T', 'got': 'TAT', 'pos': 16811082},\n",
      " {'ID': 'MU38161712', 'expected': 'TC', 'got': 'CATC', 'pos': 169086078}]\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "from collections import defaultdict\n",
    "import requests, sys\n",
    "\n",
    "\n",
    "# < --- Request server and headers\n",
    "\n",
    "grch38_server = \"http://rest.ensembl.org\"\n",
    "grch37_server = \"http://grch37.rest.ensembl.org\" # We are using the GRCh37 server\n",
    "\n",
    "ext = \"/sequence/region/human\"\n",
    "headers={ \"Content-Type\" : \"application/json\", \n",
    "          \"Accept\" : \"application/json\"}\n",
    "\n",
    "\n",
    "# < --- Validation\n",
    "\n",
    "def validate_mutations(mutations, verbose=False, assembly='GRCh37'):\n",
    "    \"Return the mutation IDs that doesn't match the reference.\"\n",
    "    invalid = []  # Here we will accumulate those that didn't match\n",
    "    \n",
    "    # < --- Assemble the request data\n",
    "    request_data = defaultdict(list)\n",
    "    for mutation in mutations.itertuples():\n",
    "        # Add region string to the data dictionary\n",
    "        # at the automatically created 'region' key\n",
    "        request_data['regions'].append( to_region_str(mutation, assembly) )\n",
    "\n",
    "    # < --- Make the request\n",
    "    server = grch37_server if assembly=='GRCh37' else grch38_server\n",
    "    \n",
    "    # Request the sequences at the position\n",
    "    r = requests.post(server+ext, \n",
    "                      headers=headers, \n",
    "                      data=json.dumps(request_data)) # Here goes the data\n",
    "\n",
    "    # < --- Check the response\n",
    "    if not r.ok:\n",
    "        r.raise_for_status()\n",
    "\n",
    "    # < --- Validate the mutations\n",
    "    for r_item, mut in zip(r.json(), mutations.itertuples()):\n",
    "        # The sequence from the file\n",
    "        file_seq = mut.MUTATION.split('>')[0]\n",
    "        # The sequence from the request\n",
    "        req_seq = r_item['seq']\n",
    "        \n",
    "        # < --- Parse the mutation's start position\n",
    "        if assembly == 'GRCh37':\n",
    "            chrom_str, pos_str = mut.POSITION_GRCh37.split(':')\n",
    "        else:\n",
    "            chrom_str, pos_str = mut.POSITION_GRCh38.split(':')\n",
    "        \n",
    "        if file_seq != req_seq[1:-1]:\n",
    "            invalid.append({'ID':mut.MUTATION_ID,\n",
    "                            'expected':file_seq,\n",
    "                            'got':req_seq,\n",
    "                            'pos':int(pos_str)})\n",
    "        if verbose:\n",
    "            print('ID: ', mut.MUTATION_ID,\n",
    "                  'Expected: ', file_seq,\n",
    "                  'Got: ', req_seq)\n",
    "        \n",
    "    return invalid\n",
    "# ---\n",
    "\n",
    "\n",
    "# Test\n",
    "\n",
    "from pprint import pprint\n",
    "\n",
    "pprint(validate_mutations(mutations, \n",
    "                          verbose=True, \n",
    "                          assembly='GRCh37'))\n",
    "pprint(validate_mutations(mutations, \n",
    "                          verbose=True, \n",
    "                          assembly='GRCh38'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Processing all mutations\n",
    "==============\n",
    "\n",
    "Now that every part of the algorithm is OK, let's parse the complete file!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations_reader = pd.read_table('~/results/all.BRCA-EU.mutations-context.tsv', \n",
    "                                 header=1, \n",
    "                                 delim_whitespace=True,\n",
    "                                 chunksize=50)  # Changed a little to ease iteration [1].\n",
    "                                                # chunksize=50 because that's the maximum\n",
    "                                                # allowed per request by Ensembl [2].\n",
    "# [1]. See: https://pandas.pydata.org/pandas-docs/stable/io.html#io-chunking\n",
    "# [2]. See: http://rest.ensembl.org/documentation/info/sequence_region_post"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk:  0 --------------------\n",
      "[{'ID': 'MU64619292', 'expected': 'T', 'got': 'TAT', 'pos': 16992349},\n",
      " {'ID': 'MU38161712', 'expected': 'TC', 'got': 'CATC', 'pos': 169942588}]\n",
      "\n",
      "\n",
      "[{'ID': 'MU64619292', 'expected': 'T', 'got': 'TAT', 'pos': 16811082},\n",
      " {'ID': 'MU38161712', 'expected': 'TC', 'got': 'CATC', 'pos': 169086078}]\n",
      "Chunk:  1 --------------------\n",
      "[{'ID': 'MU63854834', 'expected': 'A', 'got': 'GGA', 'pos': 169967250},\n",
      " {'ID': 'MU64350164', 'expected': 'T', 'got': 'TAT', 'pos': 16996995},\n",
      " {'ID': 'MU63418111', 'expected': '-', 'got': 'AGA', 'pos': 169982032},\n",
      " {'ID': 'MU64994227', 'expected': '-', 'got': 'TTA', 'pos': 169984697},\n",
      " {'ID': 'MU63952827', 'expected': 'AAGT', 'got': 'AAAAGT', 'pos': 169987319},\n",
      " {'ID': 'MU64982903', 'expected': 'TT', 'got': 'TCTT', 'pos': 169989944}]\n",
      "\n",
      "\n",
      "[{'ID': 'MU63854834', 'expected': 'A', 'got': 'GGA', 'pos': 169110740},\n",
      " {'ID': 'MU64350164', 'expected': 'T', 'got': 'TAT', 'pos': 16815728},\n",
      " {'ID': 'MU63418111', 'expected': '-', 'got': 'AGA', 'pos': 169125522},\n",
      " {'ID': 'MU64994227', 'expected': '-', 'got': 'TTA', 'pos': 169128187},\n",
      " {'ID': 'MU63952827', 'expected': 'AAGT', 'got': 'AAAAGT', 'pos': 169130809},\n",
      " {'ID': 'MU64982903', 'expected': 'TT', 'got': 'TCTT', 'pos': 169133434}]\n",
      "Chunk:  2 --------------------\n",
      "[{'ID': 'MU130883', 'expected': 'G', 'got': 'ATG', 'pos': 169997024},\n",
      " {'ID': 'MU64367222', 'expected': 'A', 'got': 'GGA', 'pos': 170005227},\n",
      " {'ID': 'MU64018271',\n",
      "  'expected': 'ATATAACAGAATT',\n",
      "  'got': 'AAATATAACAGAATT',\n",
      "  'pos': 170006654},\n",
      " {'ID': 'MU63766231', 'expected': '-', 'got': 'TTA', 'pos': 170018426}]\n",
      "\n",
      "\n",
      "[{'ID': 'MU130883', 'expected': 'G', 'got': 'ATG', 'pos': 169140514},\n",
      " {'ID': 'MU64367222', 'expected': 'A', 'got': 'GGA', 'pos': 169148717},\n",
      " {'ID': 'MU64018271',\n",
      "  'expected': 'ATATAACAGAATT',\n",
      "  'got': 'AAATATAACAGAATT',\n",
      "  'pos': 169150144},\n",
      " {'ID': 'MU63766231', 'expected': '-', 'got': 'TTA', 'pos': 169161916}]\n",
      "Chunk:  3 --------------------\n",
      "[{'ID': 'MU64002203', 'expected': 'TTAA', 'got': 'CTTTAA', 'pos': 170034676},\n",
      " {'ID': 'MU63446893', 'expected': 'A', 'got': 'GGA', 'pos': 170034775},\n",
      " {'ID': 'MU64925025', 'expected': 'A', 'got': 'AGA', 'pos': 170041878},\n",
      " {'ID': 'MU43769937', 'expected': 'T', 'got': 'CCT', 'pos': 170058056}]\n",
      "\n",
      "\n",
      "[{'ID': 'MU64002203', 'expected': 'TTAA', 'got': 'CTTTAA', 'pos': 169178166},\n",
      " {'ID': 'MU63446893', 'expected': 'A', 'got': 'GGA', 'pos': 169178265},\n",
      " {'ID': 'MU64925025', 'expected': 'A', 'got': 'AGA', 'pos': 169185368},\n",
      " {'ID': 'MU43769937', 'expected': 'T', 'got': 'CCT', 'pos': 169201546}]\n",
      "Chunk:  4 --------------------\n",
      "[{'ID': 'MU64803253', 'expected': 'T', 'got': 'GGT', 'pos': 17007427},\n",
      " {'ID': 'MU63430793', 'expected': '-', 'got': 'ACT', 'pos': 170085411},\n",
      " {'ID': 'MU64367227', 'expected': 'T', 'got': 'TGT', 'pos': 170091923}]\n",
      "\n",
      "\n",
      "[{'ID': 'MU64803253', 'expected': 'T', 'got': 'GGT', 'pos': 16826160},\n",
      " {'ID': 'MU63430793', 'expected': '-', 'got': 'ACT', 'pos': 169228901},\n",
      " {'ID': 'MU64367227', 'expected': 'T', 'got': 'TGT', 'pos': 169235413}]\n",
      "Chunk:  5 --------------------\n",
      "[{'ID': 'MU64011232', 'expected': 'A', 'got': 'GCA', 'pos': 170115539},\n",
      " {'ID': 'MU64997386', 'expected': '-', 'got': 'AGA', 'pos': 170116021},\n",
      " {'ID': 'MU64350170', 'expected': 'T', 'got': 'AAT', 'pos': 17011629}]\n",
      "\n",
      "\n",
      "[{'ID': 'MU64011232', 'expected': 'A', 'got': 'GCA', 'pos': 169259029},\n",
      " {'ID': 'MU64997386', 'expected': '-', 'got': 'AGA', 'pos': 169259511},\n",
      " {'ID': 'MU64350170', 'expected': 'T', 'got': 'AAT', 'pos': 16830362}]\n",
      "Chunk:  6 --------------------\n",
      "[{'ID': 'MU64367229', 'expected': 'G', 'got': 'GCG', 'pos': 170150578},\n",
      " {'ID': 'MU64063560', 'expected': '-', 'got': 'GGT', 'pos': 170157256},\n",
      " {'ID': 'MU63410569', 'expected': 'TATGCT', 'got': 'ACTATGCT', 'pos': 17017047}]\n",
      "\n",
      "\n",
      "[{'ID': 'MU64367229', 'expected': 'G', 'got': 'GCG', 'pos': 169294068},\n",
      " {'ID': 'MU64063560', 'expected': '-', 'got': 'GGT', 'pos': 169300746},\n",
      " {'ID': 'MU63410569', 'expected': 'TATGCT', 'got': 'ACTATGCT', 'pos': 16835780}]\n",
      "Chunk:  7 --------------------\n",
      "[{'ID': 'MU64631113', 'expected': 'A', 'got': 'TGA', 'pos': 170179286},\n",
      " {'ID': 'MU63812497', 'expected': 'T', 'got': 'CAT', 'pos': 170179697},\n",
      " {'ID': 'MU63631765', 'expected': '-', 'got': 'TCT', 'pos': 170193904},\n",
      " {'ID': 'MU65005483', 'expected': 'AAGT', 'got': 'AGAAGT', 'pos': 170202505},\n",
      " {'ID': 'MU64367232', 'expected': 'G', 'got': 'CTG', 'pos': 170208717}]\n",
      "\n",
      "\n",
      "[{'ID': 'MU64631113', 'expected': 'A', 'got': 'TGA', 'pos': 169322776},\n",
      " {'ID': 'MU63812497', 'expected': 'T', 'got': 'CAT', 'pos': 169323187},\n",
      " {'ID': 'MU63631765', 'expected': '-', 'got': 'TCT', 'pos': 169337394},\n",
      " {'ID': 'MU65005483', 'expected': 'AAGT', 'got': 'AGAAGT', 'pos': 169345995},\n",
      " {'ID': 'MU64367232', 'expected': 'G', 'got': 'CTG', 'pos': 169352207}]\n",
      "Chunk:  8 --------------------\n",
      "[{'ID': 'MU63686381', 'expected': 'T', 'got': 'TCT', 'pos': 170215957},\n",
      " {'ID': 'MU63718086', 'expected': 'C', 'got': 'CTC', 'pos': 170223014},\n",
      " {'ID': 'MU63909649', 'expected': '-', 'got': 'TTA', 'pos': 170226532},\n",
      " {'ID': 'MU64367236', 'expected': 'T', 'got': 'AAT', 'pos': 170232553},\n",
      " {'ID': 'MU64011236', 'expected': '-', 'got': 'GGA', 'pos': 170233180},\n",
      " {'ID': 'MU63927098', 'expected': '-', 'got': 'CTA', 'pos': 170236748},\n",
      " {'ID': 'MU63659587',\n",
      "  'expected': 'TTCAGGCAGTTGTAAGATTCGTATTGGC',\n",
      "  'got': 'TTTTCAGGCAGTTGTAAGATTCGTATTGGC',\n",
      "  'pos': 170246020}]\n",
      "\n",
      "\n",
      "[{'ID': 'MU63686381', 'expected': 'T', 'got': 'TCT', 'pos': 169359447},\n",
      " {'ID': 'MU63718086', 'expected': 'C', 'got': 'CTC', 'pos': 169366504},\n",
      " {'ID': 'MU63909649', 'expected': '-', 'got': 'TTA', 'pos': 169370022},\n",
      " {'ID': 'MU64367236', 'expected': 'T', 'got': 'AAT', 'pos': 169376043},\n",
      " {'ID': 'MU64011236', 'expected': '-', 'got': 'GGA', 'pos': 169376670},\n",
      " {'ID': 'MU63927098', 'expected': '-', 'got': 'CTA', 'pos': 169380238},\n",
      " {'ID': 'MU63659587',\n",
      "  'expected': 'TTCAGGCAGTTGTAAGATTCGTATTGGC',\n",
      "  'got': 'TTTTCAGGCAGTTGTAAGATTCGTATTGGC',\n",
      "  'pos': 169389510}]\n",
      "Chunk:  9 --------------------\n",
      "[]\n",
      "\n",
      "\n",
      "[]\n",
      "Chunk:  10 --------------------\n",
      "[{'ID': 'MU64631115', 'expected': 'C', 'got': 'AAC', 'pos': 170283488},\n",
      " {'ID': 'MU64619296', 'expected': 'T', 'got': 'TGT', 'pos': 17028450},\n",
      " {'ID': 'MU64631118', 'expected': 'A', 'got': 'TTA', 'pos': 170301846},\n",
      " {'ID': 'MU64631123', 'expected': '-', 'got': 'TGA', 'pos': 170302971},\n",
      " {'ID': 'MU63854836', 'expected': 'A', 'got': 'TTA', 'pos': 170302993},\n",
      " {'ID': 'MU63850061', 'expected': 'A', 'got': 'ACA', 'pos': 17030680},\n",
      " {'ID': 'MU63944640', 'expected': 'A', 'got': 'AGA', 'pos': 170309922}]\n",
      "\n",
      "\n",
      "[{'ID': 'MU64631115', 'expected': 'C', 'got': 'AAC', 'pos': 169426978},\n",
      " {'ID': 'MU64619296', 'expected': 'T', 'got': 'TGT', 'pos': 16847183},\n",
      " {'ID': 'MU64631118', 'expected': 'A', 'got': 'TTA', 'pos': 169445336},\n",
      " {'ID': 'MU64631123', 'expected': '-', 'got': 'TGA', 'pos': 169446461},\n",
      " {'ID': 'MU63854836', 'expected': 'A', 'got': 'TTA', 'pos': 169446483},\n",
      " {'ID': 'MU63850061', 'expected': 'A', 'got': 'ACA', 'pos': 16849413},\n",
      " {'ID': 'MU63944640', 'expected': 'A', 'got': 'AGA', 'pos': 169453412}]\n",
      "Chunk:  11 --------------------\n",
      "[{'ID': 'MU64367239', 'expected': 'T', 'got': 'CAT', 'pos': 170321604},\n",
      " {'ID': 'MU64367242', 'expected': 'T', 'got': 'TCT', 'pos': 170327916},\n",
      " {'ID': 'MU64168859', 'expected': 'G', 'got': 'ATG', 'pos': 170329821},\n",
      " {'ID': 'MU64631129', 'expected': 'T', 'got': 'GAT', 'pos': 170346215},\n",
      " {'ID': 'MU64808479', 'expected': 'T', 'got': 'TAT', 'pos': 170346337}]\n",
      "\n",
      "\n",
      "[{'ID': 'MU64367239', 'expected': 'T', 'got': 'CAT', 'pos': 169465094},\n",
      " {'ID': 'MU64367242', 'expected': 'T', 'got': 'TCT', 'pos': 169471406},\n",
      " {'ID': 'MU64168859', 'expected': 'G', 'got': 'ATG', 'pos': 169473311},\n",
      " {'ID': 'MU64631129', 'expected': 'T', 'got': 'GAT', 'pos': 169489705},\n",
      " {'ID': 'MU64808479', 'expected': 'T', 'got': 'TAT', 'pos': 169489827}]\n",
      "Chunk:  12 --------------------\n",
      "[{'ID': 'MU63733411',\n",
      "  'expected': 'TTCGAGA',\n",
      "  'got': 'GGTTCGAGA',\n",
      "  'pos': 1703568},\n",
      " {'ID': 'MU63906012', 'expected': 'T', 'got': 'AGT', 'pos': 170369034},\n",
      " {'ID': 'MU65018331', 'expected': 'TA', 'got': 'GGTA', 'pos': 17037520},\n",
      " {'ID': 'MU64237230', 'expected': 'T', 'got': 'CAT', 'pos': 170379307},\n",
      " {'ID': 'MU64617980', 'expected': 'T', 'got': 'TAT', 'pos': 1703842},\n",
      " {'ID': 'MU64367246', 'expected': 'T', 'got': 'GAT', 'pos': 170392129},\n",
      " {'ID': 'MU63501908', 'expected': 'AT', 'got': 'AAAT', 'pos': 170392917}]\n",
      "\n",
      "\n",
      "[{'ID': 'MU63733411',\n",
      "  'expected': 'TTCGAGA',\n",
      "  'got': 'GGTTCGAGA',\n",
      "  'pos': 1699796},\n",
      " {'ID': 'MU63906012', 'expected': 'T', 'got': 'AGT', 'pos': 169512524},\n",
      " {'ID': 'MU65018331', 'expected': 'TA', 'got': 'GGTA', 'pos': 16856253},\n",
      " {'ID': 'MU64237230', 'expected': 'T', 'got': 'CAT', 'pos': 169522797},\n",
      " {'ID': 'MU64617980', 'expected': 'T', 'got': 'TAT', 'pos': 1700070},\n",
      " {'ID': 'MU64367246', 'expected': 'T', 'got': 'GAT', 'pos': 169535619},\n",
      " {'ID': 'MU63501908', 'expected': 'AT', 'got': 'AAAT', 'pos': 169536407}]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk:  13 --------------------\n",
      "[{'ID': 'MU64367249', 'expected': 'ATT', 'got': 'AAATT', 'pos': 170420944},\n",
      " {'ID': 'MU64367252', 'expected': 'T', 'got': 'CAT', 'pos': 170421411},\n",
      " {'ID': 'MU63501914', 'expected': 'T', 'got': 'AGT', 'pos': 170434087},\n",
      " {'ID': 'MU64631133', 'expected': 'A', 'got': 'ACA', 'pos': 170434765}]\n",
      "\n",
      "\n",
      "[{'ID': 'MU64367249', 'expected': 'ATT', 'got': 'AAATT', 'pos': 169564434},\n",
      " {'ID': 'MU64367252', 'expected': 'T', 'got': 'CAT', 'pos': 169564901},\n",
      " {'ID': 'MU63501914', 'expected': 'T', 'got': 'AGT', 'pos': 169577577},\n",
      " {'ID': 'MU64631133', 'expected': 'A', 'got': 'ACA', 'pos': 169578255}]\n",
      "Chunk:  14 --------------------\n",
      "[{'ID': 'MU64631137', 'expected': 'T', 'got': 'AAT', 'pos': 170448688},\n",
      " {'ID': 'MU63913109', 'expected': '-', 'got': 'TCT', 'pos': 170449121},\n",
      " {'ID': 'MU64367256', 'expected': 'A', 'got': 'TTA', 'pos': 170455975},\n",
      " {'ID': 'MU64367259', 'expected': 'T', 'got': 'TGT', 'pos': 170456137},\n",
      " {'ID': 'MU64808483', 'expected': '-', 'got': 'TGT', 'pos': 170456137},\n",
      " {'ID': 'MU64953087', 'expected': '-', 'got': 'TGT', 'pos': 170460175}]\n",
      "\n",
      "\n",
      "[{'ID': 'MU64631137', 'expected': 'T', 'got': 'AAT', 'pos': 169592178},\n",
      " {'ID': 'MU63913109', 'expected': '-', 'got': 'TCT', 'pos': 169592611},\n",
      " {'ID': 'MU64367256', 'expected': 'A', 'got': 'TTA', 'pos': 169599465},\n",
      " {'ID': 'MU64367259', 'expected': 'T', 'got': 'TGT', 'pos': 169599627},\n",
      " {'ID': 'MU64808483', 'expected': '-', 'got': 'TGT', 'pos': 169599627},\n",
      " {'ID': 'MU64953087', 'expected': '-', 'got': 'TGT', 'pos': 169603665}]\n",
      "Chunk:  15 --------------------\n",
      "[{'ID': 'MU64367262', 'expected': '-', 'got': 'ATA', 'pos': 170480414},\n",
      " {'ID': 'MU63413244',\n",
      "  'expected': 'AGGTTAAGCTCGGAATTTGTATTACTGTCTACTAGATC',\n",
      "  'got': 'CAAGGTTAAGCTCGGAATTTGTATTACTGTCTACTAGATC',\n",
      "  'pos': 170498181},\n",
      " {'ID': 'MU64367266', 'expected': 'T', 'got': 'AGT', 'pos': 170516526},\n",
      " {'ID': 'MU64367271', 'expected': '-', 'got': 'TAT', 'pos': 170517286}]\n",
      "\n",
      "\n",
      "[{'ID': 'MU64367262', 'expected': '-', 'got': 'ATA', 'pos': 169623904},\n",
      " {'ID': 'MU63413244',\n",
      "  'expected': 'AGGTTAAGCTCGGAATTTGTATTACTGTCTACTAGATC',\n",
      "  'got': 'CAAGGTTAAGCTCGGAATTTGTATTACTGTCTACTAGATC',\n",
      "  'pos': 169641671},\n",
      " {'ID': 'MU64367266', 'expected': 'T', 'got': 'AGT', 'pos': 169660016},\n",
      " {'ID': 'MU64367271', 'expected': '-', 'got': 'TAT', 'pos': 169660776}]\n",
      "Chunk:  16 --------------------\n",
      "[{'ID': 'MU64367275', 'expected': 'AAAC', 'got': 'AAAAAC', 'pos': 170522858},\n",
      " {'ID': 'MU64631140', 'expected': 'A', 'got': 'TTA', 'pos': 170530342},\n",
      " {'ID': 'MU63501917', 'expected': 'A', 'got': 'TTA', 'pos': 170530401},\n",
      " {'ID': 'MU64044998', 'expected': 'C', 'got': 'AAC', 'pos': 17053520},\n",
      " {'ID': 'MU64631144', 'expected': 'A', 'got': 'GTA', 'pos': 170537755},\n",
      " {'ID': 'MU64619301', 'expected': 'G', 'got': 'ATG', 'pos': 17054047},\n",
      " {'ID': 'MU63442611', 'expected': '-', 'got': 'TTG', 'pos': 170547969},\n",
      " {'ID': 'MU63594297', 'expected': '-', 'got': 'CTG', 'pos': 170560048},\n",
      " {'ID': 'MU64808487', 'expected': 'A', 'got': 'AGA', 'pos': 170561722}]\n",
      "\n",
      "\n",
      "[{'ID': 'MU64367275', 'expected': 'AAAC', 'got': 'AAAAAC', 'pos': 169666348},\n",
      " {'ID': 'MU64631140', 'expected': 'A', 'got': 'TTA', 'pos': 169673832},\n",
      " {'ID': 'MU63501917', 'expected': 'A', 'got': 'TTA', 'pos': 169673891},\n",
      " {'ID': 'MU64044998', 'expected': 'C', 'got': 'AAC', 'pos': 16872253},\n",
      " {'ID': 'MU64631144', 'expected': 'A', 'got': 'GTA', 'pos': 169681245},\n",
      " {'ID': 'MU64619301', 'expected': 'G', 'got': 'ATG', 'pos': 16872780},\n",
      " {'ID': 'MU63442611', 'expected': '-', 'got': 'TTG', 'pos': 169691459},\n",
      " {'ID': 'MU63594297', 'expected': '-', 'got': 'CTG', 'pos': 169703538},\n",
      " {'ID': 'MU64808487', 'expected': 'A', 'got': 'AGA', 'pos': 169705212}]\n",
      "Chunk:  17 --------------------\n",
      "[{'ID': 'MU64808490', 'expected': 'C', 'got': 'CTC', 'pos': 170576214},\n",
      " {'ID': 'MU64631147', 'expected': 'T', 'got': 'TCT', 'pos': 170577411},\n",
      " {'ID': 'MU64128490', 'expected': 'G', 'got': 'GAG', 'pos': 170581158},\n",
      " {'ID': 'MU64367278', 'expected': 'T', 'got': 'ACT', 'pos': 170585253},\n",
      " {'ID': 'MU63854840', 'expected': 'T', 'got': 'TAT', 'pos': 170601597},\n",
      " {'ID': 'MU64367283', 'expected': '-', 'got': 'TTA', 'pos': 170605896},\n",
      " {'ID': 'MU64348865', 'expected': 'C', 'got': 'CAC', 'pos': 1706070}]\n",
      "\n",
      "\n",
      "[{'ID': 'MU64808490', 'expected': 'C', 'got': 'CTC', 'pos': 169719704},\n",
      " {'ID': 'MU64631147', 'expected': 'T', 'got': 'TCT', 'pos': 169720901},\n",
      " {'ID': 'MU64128490', 'expected': 'G', 'got': 'GAG', 'pos': 169724648},\n",
      " {'ID': 'MU64367278', 'expected': 'T', 'got': 'ACT', 'pos': 169728743},\n",
      " {'ID': 'MU63854840', 'expected': 'T', 'got': 'TAT', 'pos': 169745087},\n",
      " {'ID': 'MU64367283', 'expected': '-', 'got': 'TTA', 'pos': 169749386},\n",
      " {'ID': 'MU64348865', 'expected': 'C', 'got': 'CAC', 'pos': 1702298}]\n"
     ]
    },
    {
     "ename": "HTTPError",
     "evalue": "429 Client Error: Too Many Requests for url: http://grch37.rest.ensembl.org/sequence/region/human",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mHTTPError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-122-1864c2bf85f0>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmutation_chunk\u001b[0m \u001b[0;32min\u001b[0m \u001b[0menumerate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmutations_reader\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m     \u001b[0;31m# Validate the chunk\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m     \u001b[0mnew_invalid_grch37\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mvalidate_mutations\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmutation_chunk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0massembly\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'GRCh37'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     10\u001b[0m     \u001b[0mnew_invalid_grch38\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mvalidate_mutations\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmutation_chunk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0massembly\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'GRCh38'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Chunk: '\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'-'\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0;36m20\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-120-0f90322ec660>\u001b[0m in \u001b[0;36mvalidate_mutations\u001b[0;34m(mutations, verbose, assembly)\u001b[0m\n\u001b[1;32m     37\u001b[0m     \u001b[0;31m# < --- Check the response\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     38\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mok\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 39\u001b[0;31m         \u001b[0mr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mraise_for_status\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     40\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     41\u001b[0m     \u001b[0;31m# < --- Validate the mutations\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.6/site-packages/requests/models.py\u001b[0m in \u001b[0;36mraise_for_status\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    933\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    934\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mhttp_error_msg\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 935\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mHTTPError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhttp_error_msg\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mresponse\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    936\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    937\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mHTTPError\u001b[0m: 429 Client Error: Too Many Requests for url: http://grch37.rest.ensembl.org/sequence/region/human"
     ]
    }
   ],
   "source": [
    "from pprint import pprint\n",
    "\n",
    "invalid_grch37 = []\n",
    "invalid_grch38 = []\n",
    "\n",
    "# Iterate through each chunk\n",
    "for i, mutation_chunk in enumerate(mutations_reader):\n",
    "    # Validate the chunk\n",
    "    new_invalid_grch37 = validate_mutations(mutation_chunk, assembly='GRCh37')\n",
    "    new_invalid_grch38 = validate_mutations(mutation_chunk, assembly='GRCh38')\n",
    "    print('Chunk: ', i, '-'*20)\n",
    "    pprint(new_invalid_grch37)\n",
    "    print('\\n')\n",
    "    pprint(new_invalid_grch38)\n",
    "    \n",
    "    invalid_grch37 += new_invalid_grch37\n",
    "    invalid_grch38 += new_invalid_grch38\n",
    "\n",
    "pprint(invalid_grch37)\n",
    "pprint(invalid_grch38)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ok, looks like we saturated GRCh37 server, so let's ask for only the GRCh38 mutations..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pprint import pprint\n",
    "\n",
    "invalid_grch38 = []\n",
    "\n",
    "# Iterate through each chunk\n",
    "for i, mutation_chunk in enumerate(mutations_reader):\n",
    "    # Validate the chunk\n",
    "    new_invalid_grch38 = validate_mutations(mutation_chunk, assembly='GRCh38')\n",
    "    print('Chunk: ', i, '-'*20)\n",
    "    pprint(new_invalid_grch38)\n",
    "    invalid_grch38 += new_invalid_grch38\n",
    "\n",
    "pprint(invalid_grch38)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Validating the original file\n",
    "================\n",
    "\n",
    "As the mutations are wrong, no matter what assembly we are talking about, the next step is to validate things directly from source. The ICGC's Simple Somatic Mutations file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ID': 'MU64619292', 'expected': 'T', 'got': 'TAT', 'pos': 16992349}"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get one of the invalidated mutations\n",
    "m_not_valid = invalid_grch37[0]\n",
    "m_not_valid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ssm_alt.sqlite\tssm.sqlite  ssm.vcf\r\n"
     ]
    }
   ],
   "source": [
    "## Searching the data file\n",
    "! ls ../data/data-release-25/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The file `ssm.sqlite` is a pre-made database file containing the mutations. Let's inspect it..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "import dataset\n",
    "\n",
    "# Connect to the database\n",
    "database = dataset.connect('sqlite:///../data/data-release-25/ssm.sqlite')\n",
    "\n",
    "# Connect to the mutations table\n",
    "mut_table = database['Mutation']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Consequence',\n",
       " 'Consequence_Mutation',\n",
       " 'Mutation',\n",
       " 'Mutation_OccurrenceByProject',\n",
       " 'OccurrenceByProject',\n",
       " 'OccurrenceGlobal',\n",
       " 'sqlite_sequence']"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "database.tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['id',\n",
       " 'mutation_id',\n",
       " 'chromosome',\n",
       " 'GRCh37_pos',\n",
       " 'reference_allele',\n",
       " 'mutated_allele',\n",
       " 'quality',\n",
       " 'filter',\n",
       " 'occurrence_global']"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut_table.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "OrderedDict([('id', 27323401),\n",
       "             ('mutation_id', 'MU64619292'),\n",
       "             ('chromosome', '2'),\n",
       "             ('GRCh37_pos', 16992349),\n",
       "             ('reference_allele', 'AT'),\n",
       "             ('mutated_allele', 'A'),\n",
       "             ('quality', '.'),\n",
       "             ('filter', '.'),\n",
       "             ('occurrence_global', 27323401)])"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut_table.find_one(mutation_id='MU64619292')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ID': 'MU64619292', 'expected': 'T', 'got': 'TAT', 'pos': 16992349}"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# To compare with the db result\n",
    "m_not_valid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Compare the positions\n",
    "\n",
    "# To accumulate the not matching ones\n",
    "not_matching = list()\n",
    "\n",
    "for invalid in invalid_grch37:\n",
    "    # Get the original mutation record\n",
    "    original = mut_table.find_one(mutation_id=invalid['ID'])\n",
    "    # Fetch the position from there\n",
    "    original_position = original['GRCh37_pos']\n",
    "    original_mutation = original['reference_allele']\n",
    "    # Check if the positions match\n",
    "    if original_position != invalid['pos']:\n",
    "        not_matching.append({'ID': invalid['ID'],\n",
    "                             'pos':invalid['pos'],\n",
    "                             'original_pos':original_position,\n",
    "                             'expected':invalid['expected'],\n",
    "                             'original_expected':original_mutation,\n",
    "                             'got':invalid['got']})\n",
    "    \n",
    "not_matching"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So, the positions correspond well, let's verify the 'expected' nucleotides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'ID': 'MU64619292',\n",
       "  'db_expected': 'AT',\n",
       "  'db_pos': 16992349,\n",
       "  'expected': 'T',\n",
       "  'got': 'TAT',\n",
       "  'pos': 16992349},\n",
       " {'ID': 'MU38161712',\n",
       "  'db_expected': 'ATC',\n",
       "  'db_pos': 169942588,\n",
       "  'expected': 'TC',\n",
       "  'got': 'CATC',\n",
       "  'pos': 169942588},\n",
       " {'ID': 'MU63854834',\n",
       "  'db_expected': 'GA',\n",
       "  'db_pos': 169967250,\n",
       "  'expected': 'A',\n",
       "  'got': 'GGA',\n",
       "  'pos': 169967250},\n",
       " {'ID': 'MU64350164',\n",
       "  'db_expected': 'AT',\n",
       "  'db_pos': 16996995,\n",
       "  'expected': 'T',\n",
       "  'got': 'TAT',\n",
       "  'pos': 16996995},\n",
       " {'ID': 'MU63418111',\n",
       "  'db_expected': 'G',\n",
       "  'db_pos': 169982032,\n",
       "  'expected': '-',\n",
       "  'got': 'AGA',\n",
       "  'pos': 169982032},\n",
       " {'ID': 'MU64994227',\n",
       "  'db_expected': 'T',\n",
       "  'db_pos': 169984697,\n",
       "  'expected': '-',\n",
       "  'got': 'TTA',\n",
       "  'pos': 169984697},\n",
       " {'ID': 'MU63952827',\n",
       "  'db_expected': 'AAAGT',\n",
       "  'db_pos': 169987319,\n",
       "  'expected': 'AAGT',\n",
       "  'got': 'AAAAGT',\n",
       "  'pos': 169987319},\n",
       " {'ID': 'MU64982903',\n",
       "  'db_expected': 'CTT',\n",
       "  'db_pos': 169989944,\n",
       "  'expected': 'TT',\n",
       "  'got': 'TCTT',\n",
       "  'pos': 169989944},\n",
       " {'ID': 'MU130883',\n",
       "  'db_expected': 'TG',\n",
       "  'db_pos': 169997024,\n",
       "  'expected': 'G',\n",
       "  'got': 'ATG',\n",
       "  'pos': 169997024},\n",
       " {'ID': 'MU64367222',\n",
       "  'db_expected': 'GA',\n",
       "  'db_pos': 170005227,\n",
       "  'expected': 'A',\n",
       "  'got': 'GGA',\n",
       "  'pos': 170005227},\n",
       " {'ID': 'MU64018271',\n",
       "  'db_expected': 'AATATAACAGAATT',\n",
       "  'db_pos': 170006654,\n",
       "  'expected': 'ATATAACAGAATT',\n",
       "  'got': 'AAATATAACAGAATT',\n",
       "  'pos': 170006654},\n",
       " {'ID': 'MU63766231',\n",
       "  'db_expected': 'T',\n",
       "  'db_pos': 170018426,\n",
       "  'expected': '-',\n",
       "  'got': 'TTA',\n",
       "  'pos': 170018426},\n",
       " {'ID': 'MU64002203',\n",
       "  'db_expected': 'TTTAA',\n",
       "  'db_pos': 170034676,\n",
       "  'expected': 'TTAA',\n",
       "  'got': 'CTTTAA',\n",
       "  'pos': 170034676},\n",
       " {'ID': 'MU63446893',\n",
       "  'db_expected': 'GA',\n",
       "  'db_pos': 170034775,\n",
       "  'expected': 'A',\n",
       "  'got': 'GGA',\n",
       "  'pos': 170034775},\n",
       " {'ID': 'MU64925025',\n",
       "  'db_expected': 'GA',\n",
       "  'db_pos': 170041878,\n",
       "  'expected': 'A',\n",
       "  'got': 'AGA',\n",
       "  'pos': 170041878},\n",
       " {'ID': 'MU43769937',\n",
       "  'db_expected': 'CT',\n",
       "  'db_pos': 170058056,\n",
       "  'expected': 'T',\n",
       "  'got': 'CCT',\n",
       "  'pos': 170058056},\n",
       " {'ID': 'MU64803253',\n",
       "  'db_expected': 'GT',\n",
       "  'db_pos': 17007427,\n",
       "  'expected': 'T',\n",
       "  'got': 'GGT',\n",
       "  'pos': 17007427},\n",
       " {'ID': 'MU63430793',\n",
       "  'db_expected': 'C',\n",
       "  'db_pos': 170085411,\n",
       "  'expected': '-',\n",
       "  'got': 'ACT',\n",
       "  'pos': 170085411},\n",
       " {'ID': 'MU64367227',\n",
       "  'db_expected': 'GT',\n",
       "  'db_pos': 170091923,\n",
       "  'expected': 'T',\n",
       "  'got': 'TGT',\n",
       "  'pos': 170091923},\n",
       " {'ID': 'MU64011232',\n",
       "  'db_expected': 'CA',\n",
       "  'db_pos': 170115539,\n",
       "  'expected': 'A',\n",
       "  'got': 'GCA',\n",
       "  'pos': 170115539},\n",
       " {'ID': 'MU64997386',\n",
       "  'db_expected': 'G',\n",
       "  'db_pos': 170116021,\n",
       "  'expected': '-',\n",
       "  'got': 'AGA',\n",
       "  'pos': 170116021},\n",
       " {'ID': 'MU64350170',\n",
       "  'db_expected': 'AT',\n",
       "  'db_pos': 17011629,\n",
       "  'expected': 'T',\n",
       "  'got': 'AAT',\n",
       "  'pos': 17011629},\n",
       " {'ID': 'MU64367229',\n",
       "  'db_expected': 'CG',\n",
       "  'db_pos': 170150578,\n",
       "  'expected': 'G',\n",
       "  'got': 'GCG',\n",
       "  'pos': 170150578},\n",
       " {'ID': 'MU64063560',\n",
       "  'db_expected': 'G',\n",
       "  'db_pos': 170157256,\n",
       "  'expected': '-',\n",
       "  'got': 'GGT',\n",
       "  'pos': 170157256},\n",
       " {'ID': 'MU63410569',\n",
       "  'db_expected': 'CTATGCT',\n",
       "  'db_pos': 17017047,\n",
       "  'expected': 'TATGCT',\n",
       "  'got': 'ACTATGCT',\n",
       "  'pos': 17017047},\n",
       " {'ID': 'MU64631113',\n",
       "  'db_expected': 'GA',\n",
       "  'db_pos': 170179286,\n",
       "  'expected': 'A',\n",
       "  'got': 'TGA',\n",
       "  'pos': 170179286},\n",
       " {'ID': 'MU63812497',\n",
       "  'db_expected': 'AT',\n",
       "  'db_pos': 170179697,\n",
       "  'expected': 'T',\n",
       "  'got': 'CAT',\n",
       "  'pos': 170179697},\n",
       " {'ID': 'MU63631765',\n",
       "  'db_expected': 'C',\n",
       "  'db_pos': 170193904,\n",
       "  'expected': '-',\n",
       "  'got': 'TCT',\n",
       "  'pos': 170193904},\n",
       " {'ID': 'MU65005483',\n",
       "  'db_expected': 'GAAGT',\n",
       "  'db_pos': 170202505,\n",
       "  'expected': 'AAGT',\n",
       "  'got': 'AGAAGT',\n",
       "  'pos': 170202505},\n",
       " {'ID': 'MU64367232',\n",
       "  'db_expected': 'TG',\n",
       "  'db_pos': 170208717,\n",
       "  'expected': 'G',\n",
       "  'got': 'CTG',\n",
       "  'pos': 170208717},\n",
       " {'ID': 'MU63686381',\n",
       "  'db_expected': 'CT',\n",
       "  'db_pos': 170215957,\n",
       "  'expected': 'T',\n",
       "  'got': 'TCT',\n",
       "  'pos': 170215957},\n",
       " {'ID': 'MU63718086',\n",
       "  'db_expected': 'TC',\n",
       "  'db_pos': 170223014,\n",
       "  'expected': 'C',\n",
       "  'got': 'CTC',\n",
       "  'pos': 170223014},\n",
       " {'ID': 'MU63909649',\n",
       "  'db_expected': 'T',\n",
       "  'db_pos': 170226532,\n",
       "  'expected': '-',\n",
       "  'got': 'TTA',\n",
       "  'pos': 170226532},\n",
       " {'ID': 'MU64367236',\n",
       "  'db_expected': 'AT',\n",
       "  'db_pos': 170232553,\n",
       "  'expected': 'T',\n",
       "  'got': 'AAT',\n",
       "  'pos': 170232553},\n",
       " {'ID': 'MU64011236',\n",
       "  'db_expected': 'G',\n",
       "  'db_pos': 170233180,\n",
       "  'expected': '-',\n",
       "  'got': 'GGA',\n",
       "  'pos': 170233180},\n",
       " {'ID': 'MU63927098',\n",
       "  'db_expected': 'T',\n",
       "  'db_pos': 170236748,\n",
       "  'expected': '-',\n",
       "  'got': 'CTA',\n",
       "  'pos': 170236748},\n",
       " {'ID': 'MU63659587',\n",
       "  'db_expected': 'TTTCAGGCAGTTGTAAGATTCGTATTGGC',\n",
       "  'db_pos': 170246020,\n",
       "  'expected': 'TTCAGGCAGTTGTAAGATTCGTATTGGC',\n",
       "  'got': 'TTTTCAGGCAGTTGTAAGATTCGTATTGGC',\n",
       "  'pos': 170246020},\n",
       " {'ID': 'MU64631115',\n",
       "  'db_expected': 'AC',\n",
       "  'db_pos': 170283488,\n",
       "  'expected': 'C',\n",
       "  'got': 'AAC',\n",
       "  'pos': 170283488},\n",
       " {'ID': 'MU64619296',\n",
       "  'db_expected': 'GT',\n",
       "  'db_pos': 17028450,\n",
       "  'expected': 'T',\n",
       "  'got': 'TGT',\n",
       "  'pos': 17028450},\n",
       " {'ID': 'MU64631118',\n",
       "  'db_expected': 'TA',\n",
       "  'db_pos': 170301846,\n",
       "  'expected': 'A',\n",
       "  'got': 'TTA',\n",
       "  'pos': 170301846},\n",
       " {'ID': 'MU64631123',\n",
       "  'db_expected': 'G',\n",
       "  'db_pos': 170302971,\n",
       "  'expected': '-',\n",
       "  'got': 'TGA',\n",
       "  'pos': 170302971},\n",
       " {'ID': 'MU63854836',\n",
       "  'db_expected': 'TA',\n",
       "  'db_pos': 170302993,\n",
       "  'expected': 'A',\n",
       "  'got': 'TTA',\n",
       "  'pos': 170302993},\n",
       " {'ID': 'MU63850061',\n",
       "  'db_expected': 'CA',\n",
       "  'db_pos': 17030680,\n",
       "  'expected': 'A',\n",
       "  'got': 'ACA',\n",
       "  'pos': 17030680},\n",
       " {'ID': 'MU63944640',\n",
       "  'db_expected': 'GA',\n",
       "  'db_pos': 170309922,\n",
       "  'expected': 'A',\n",
       "  'got': 'AGA',\n",
       "  'pos': 170309922},\n",
       " {'ID': 'MU64367239',\n",
       "  'db_expected': 'AT',\n",
       "  'db_pos': 170321604,\n",
       "  'expected': 'T',\n",
       "  'got': 'CAT',\n",
       "  'pos': 170321604},\n",
       " {'ID': 'MU64367242',\n",
       "  'db_expected': 'CT',\n",
       "  'db_pos': 170327916,\n",
       "  'expected': 'T',\n",
       "  'got': 'TCT',\n",
       "  'pos': 170327916},\n",
       " {'ID': 'MU64168859',\n",
       "  'db_expected': 'TG',\n",
       "  'db_pos': 170329821,\n",
       "  'expected': 'G',\n",
       "  'got': 'ATG',\n",
       "  'pos': 170329821},\n",
       " {'ID': 'MU64631129',\n",
       "  'db_expected': 'AT',\n",
       "  'db_pos': 170346215,\n",
       "  'expected': 'T',\n",
       "  'got': 'GAT',\n",
       "  'pos': 170346215},\n",
       " {'ID': 'MU64808479',\n",
       "  'db_expected': 'AT',\n",
       "  'db_pos': 170346337,\n",
       "  'expected': 'T',\n",
       "  'got': 'TAT',\n",
       "  'pos': 170346337},\n",
       " {'ID': 'MU63733411',\n",
       "  'db_expected': 'GTTCGAGA',\n",
       "  'db_pos': 1703568,\n",
       "  'expected': 'TTCGAGA',\n",
       "  'got': 'GGTTCGAGA',\n",
       "  'pos': 1703568},\n",
       " {'ID': 'MU63906012',\n",
       "  'db_expected': 'GT',\n",
       "  'db_pos': 170369034,\n",
       "  'expected': 'T',\n",
       "  'got': 'AGT',\n",
       "  'pos': 170369034},\n",
       " {'ID': 'MU65018331',\n",
       "  'db_expected': 'GTA',\n",
       "  'db_pos': 17037520,\n",
       "  'expected': 'TA',\n",
       "  'got': 'GGTA',\n",
       "  'pos': 17037520},\n",
       " {'ID': 'MU64237230',\n",
       "  'db_expected': 'AT',\n",
       "  'db_pos': 170379307,\n",
       "  'expected': 'T',\n",
       "  'got': 'CAT',\n",
       "  'pos': 170379307},\n",
       " {'ID': 'MU64617980',\n",
       "  'db_expected': 'AT',\n",
       "  'db_pos': 1703842,\n",
       "  'expected': 'T',\n",
       "  'got': 'TAT',\n",
       "  'pos': 1703842},\n",
       " {'ID': 'MU64367246',\n",
       "  'db_expected': 'AT',\n",
       "  'db_pos': 170392129,\n",
       "  'expected': 'T',\n",
       "  'got': 'GAT',\n",
       "  'pos': 170392129},\n",
       " {'ID': 'MU63501908',\n",
       "  'db_expected': 'AAT',\n",
       "  'db_pos': 170392917,\n",
       "  'expected': 'AT',\n",
       "  'got': 'AAAT',\n",
       "  'pos': 170392917},\n",
       " {'ID': 'MU64367249',\n",
       "  'db_expected': 'AATT',\n",
       "  'db_pos': 170420944,\n",
       "  'expected': 'ATT',\n",
       "  'got': 'AAATT',\n",
       "  'pos': 170420944},\n",
       " {'ID': 'MU64367252',\n",
       "  'db_expected': 'AT',\n",
       "  'db_pos': 170421411,\n",
       "  'expected': 'T',\n",
       "  'got': 'CAT',\n",
       "  'pos': 170421411},\n",
       " {'ID': 'MU63501914',\n",
       "  'db_expected': 'GT',\n",
       "  'db_pos': 170434087,\n",
       "  'expected': 'T',\n",
       "  'got': 'AGT',\n",
       "  'pos': 170434087},\n",
       " {'ID': 'MU64631133',\n",
       "  'db_expected': 'CA',\n",
       "  'db_pos': 170434765,\n",
       "  'expected': 'A',\n",
       "  'got': 'ACA',\n",
       "  'pos': 170434765},\n",
       " {'ID': 'MU64631137',\n",
       "  'db_expected': 'AT',\n",
       "  'db_pos': 170448688,\n",
       "  'expected': 'T',\n",
       "  'got': 'AAT',\n",
       "  'pos': 170448688},\n",
       " {'ID': 'MU63913109',\n",
       "  'db_expected': 'C',\n",
       "  'db_pos': 170449121,\n",
       "  'expected': '-',\n",
       "  'got': 'TCT',\n",
       "  'pos': 170449121},\n",
       " {'ID': 'MU64367256',\n",
       "  'db_expected': 'TA',\n",
       "  'db_pos': 170455975,\n",
       "  'expected': 'A',\n",
       "  'got': 'TTA',\n",
       "  'pos': 170455975},\n",
       " {'ID': 'MU64367259',\n",
       "  'db_expected': 'GT',\n",
       "  'db_pos': 170456137,\n",
       "  'expected': 'T',\n",
       "  'got': 'TGT',\n",
       "  'pos': 170456137},\n",
       " {'ID': 'MU64808483',\n",
       "  'db_expected': 'G',\n",
       "  'db_pos': 170456137,\n",
       "  'expected': '-',\n",
       "  'got': 'TGT',\n",
       "  'pos': 170456137},\n",
       " {'ID': 'MU64953087',\n",
       "  'db_expected': 'G',\n",
       "  'db_pos': 170460175,\n",
       "  'expected': '-',\n",
       "  'got': 'TGT',\n",
       "  'pos': 170460175},\n",
       " {'ID': 'MU64367262',\n",
       "  'db_expected': 'T',\n",
       "  'db_pos': 170480414,\n",
       "  'expected': '-',\n",
       "  'got': 'ATA',\n",
       "  'pos': 170480414},\n",
       " {'ID': 'MU63413244',\n",
       "  'db_expected': 'AAGGTTAAGCTCGGAATTTGTATTACTGTCTACTAGATC',\n",
       "  'db_pos': 170498181,\n",
       "  'expected': 'AGGTTAAGCTCGGAATTTGTATTACTGTCTACTAGATC',\n",
       "  'got': 'CAAGGTTAAGCTCGGAATTTGTATTACTGTCTACTAGATC',\n",
       "  'pos': 170498181},\n",
       " {'ID': 'MU64367266',\n",
       "  'db_expected': 'GT',\n",
       "  'db_pos': 170516526,\n",
       "  'expected': 'T',\n",
       "  'got': 'AGT',\n",
       "  'pos': 170516526},\n",
       " {'ID': 'MU64367271',\n",
       "  'db_expected': 'A',\n",
       "  'db_pos': 170517286,\n",
       "  'expected': '-',\n",
       "  'got': 'TAT',\n",
       "  'pos': 170517286},\n",
       " {'ID': 'MU64367275',\n",
       "  'db_expected': 'AAAAC',\n",
       "  'db_pos': 170522858,\n",
       "  'expected': 'AAAC',\n",
       "  'got': 'AAAAAC',\n",
       "  'pos': 170522858},\n",
       " {'ID': 'MU64631140',\n",
       "  'db_expected': 'TA',\n",
       "  'db_pos': 170530342,\n",
       "  'expected': 'A',\n",
       "  'got': 'TTA',\n",
       "  'pos': 170530342},\n",
       " {'ID': 'MU63501917',\n",
       "  'db_expected': 'TA',\n",
       "  'db_pos': 170530401,\n",
       "  'expected': 'A',\n",
       "  'got': 'TTA',\n",
       "  'pos': 170530401},\n",
       " {'ID': 'MU64044998',\n",
       "  'db_expected': 'AC',\n",
       "  'db_pos': 17053520,\n",
       "  'expected': 'C',\n",
       "  'got': 'AAC',\n",
       "  'pos': 17053520},\n",
       " {'ID': 'MU64631144',\n",
       "  'db_expected': 'TA',\n",
       "  'db_pos': 170537755,\n",
       "  'expected': 'A',\n",
       "  'got': 'GTA',\n",
       "  'pos': 170537755},\n",
       " {'ID': 'MU64619301',\n",
       "  'db_expected': 'TG',\n",
       "  'db_pos': 17054047,\n",
       "  'expected': 'G',\n",
       "  'got': 'ATG',\n",
       "  'pos': 17054047},\n",
       " {'ID': 'MU63442611',\n",
       "  'db_expected': 'T',\n",
       "  'db_pos': 170547969,\n",
       "  'expected': '-',\n",
       "  'got': 'TTG',\n",
       "  'pos': 170547969},\n",
       " {'ID': 'MU63594297',\n",
       "  'db_expected': 'T',\n",
       "  'db_pos': 170560048,\n",
       "  'expected': '-',\n",
       "  'got': 'CTG',\n",
       "  'pos': 170560048},\n",
       " {'ID': 'MU64808487',\n",
       "  'db_expected': 'GA',\n",
       "  'db_pos': 170561722,\n",
       "  'expected': 'A',\n",
       "  'got': 'AGA',\n",
       "  'pos': 170561722},\n",
       " {'ID': 'MU64808490',\n",
       "  'db_expected': 'TC',\n",
       "  'db_pos': 170576214,\n",
       "  'expected': 'C',\n",
       "  'got': 'CTC',\n",
       "  'pos': 170576214},\n",
       " {'ID': 'MU64631147',\n",
       "  'db_expected': 'CT',\n",
       "  'db_pos': 170577411,\n",
       "  'expected': 'T',\n",
       "  'got': 'TCT',\n",
       "  'pos': 170577411},\n",
       " {'ID': 'MU64128490',\n",
       "  'db_expected': 'AG',\n",
       "  'db_pos': 170581158,\n",
       "  'expected': 'G',\n",
       "  'got': 'GAG',\n",
       "  'pos': 170581158},\n",
       " {'ID': 'MU64367278',\n",
       "  'db_expected': 'CT',\n",
       "  'db_pos': 170585253,\n",
       "  'expected': 'T',\n",
       "  'got': 'ACT',\n",
       "  'pos': 170585253},\n",
       " {'ID': 'MU63854840',\n",
       "  'db_expected': 'AT',\n",
       "  'db_pos': 170601597,\n",
       "  'expected': 'T',\n",
       "  'got': 'TAT',\n",
       "  'pos': 170601597},\n",
       " {'ID': 'MU64367283',\n",
       "  'db_expected': 'T',\n",
       "  'db_pos': 170605896,\n",
       "  'expected': '-',\n",
       "  'got': 'TTA',\n",
       "  'pos': 170605896},\n",
       " {'ID': 'MU64348865',\n",
       "  'db_expected': 'AC',\n",
       "  'db_pos': 1706070,\n",
       "  'expected': 'C',\n",
       "  'got': 'CAC',\n",
       "  'pos': 1706070}]"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Compare the reference allele\n",
    "\n",
    "# To accumulate the not matching ones\n",
    "not_matching = list()\n",
    "\n",
    "for invalid in invalid_grch37:\n",
    "    # Get the original mutation record\n",
    "    db_mut = mut_table.find_one(mutation_id=invalid['ID'])\n",
    "    # Fetch the position from there\n",
    "    db_position = db_mut['GRCh37_pos']\n",
    "    db_ref_allele = db_mut['reference_allele']\n",
    "    # Check if the positions match\n",
    "    if db_ref_allele != invalid['expected'][1:-1]:\n",
    "        not_matching.append({'ID': invalid['ID'],\n",
    "                             'pos':invalid['pos'],\n",
    "                             'db_pos':db_position,\n",
    "                             'expected':invalid['expected'],\n",
    "                             'db_expected':db_ref_allele,\n",
    "                             'got':invalid['got']})\n",
    "    \n",
    "not_matching"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "86 86\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1.0"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(len(not_matching), len(invalid_grch37))\n",
    "len(not_matching) / len(invalid_grch37)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see that every mutation that failed to match with the Ensembl reference also failed to match with the database reference. This may mean that the `reference_allele` field was corrupted during the analysis, or, taking into account that the database was setup with the SSM file from the 25th ICGC's Data Release and the analysis was done with the equivalent file from the 22th Data Release, maybe the reference changed from a data release to another.\n",
    "\n",
    "To test this, we compare the raw files from which the analysis was made, from which the database was deployed and the final mutations file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ID': 'MU64619292', 'expected': 'T', 'got': 'TAT', 'pos': 16992349}"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## The mutation once again??\n",
    "m_not_valid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\r\n"
     ]
    }
   ],
   "source": [
    "## Print the header line\n",
    "! grep -P \"^#[^#]\" ../data/data-release-25/ssm.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\t16992349\tMU64619292\tAT\tA\t.\t.\tCONSEQUENCE=||||||intergenic_region||;OCCURRENCE=BRCA-EU|1|560|0.00179;affected_donors=1;mutation=T>-;project_count=1;tested_donors=10638\n",
      "2\t16992349\tMU64619292\tAT\tA\t.\t.\tCONSEQUENCE=||||||intergenic_region||;OCCURRENCE=BRCA-EU|1|569|0.00176;affected_donors=1;mutation=T>-;project_count=1;studies=.;tested_donors=12198\n",
      "MU64619292\tT>-\tchr2:16992349\tchr2:16811082\tINTERGENIC\t\t\tBRCA-EU\t\n"
     ]
    }
   ],
   "source": [
    "## From data release 22\n",
    "_ = ! grep MU64619292 ../data/data-release-22/ssm.vcf\n",
    "print(_[0])\n",
    "\n",
    "## From data release 25\n",
    "_ = ! grep MU64619292 ../data/data-release-25/ssm.vcf\n",
    "print(_[0])\n",
    "\n",
    "## From the analysis file\n",
    "_ = ! grep MU64619292 ~/results/all.BRCA-EU.mutations-context.tsv\n",
    "print(_[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can now clearly see that the analysis results file has the mutation string wrong! Just in case, we may try with another problematic mutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>expected</th>\n",
       "      <th>got</th>\n",
       "      <th>pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MU64619292</td>\n",
       "      <td>T</td>\n",
       "      <td>TAT</td>\n",
       "      <td>16992349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MU38161712</td>\n",
       "      <td>TC</td>\n",
       "      <td>CATC</td>\n",
       "      <td>169942588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MU63854834</td>\n",
       "      <td>A</td>\n",
       "      <td>GGA</td>\n",
       "      <td>169967250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MU64350164</td>\n",
       "      <td>T</td>\n",
       "      <td>TAT</td>\n",
       "      <td>16996995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MU63418111</td>\n",
       "      <td>-</td>\n",
       "      <td>AGA</td>\n",
       "      <td>169982032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MU64994227</td>\n",
       "      <td>-</td>\n",
       "      <td>TTA</td>\n",
       "      <td>169984697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MU63952827</td>\n",
       "      <td>AAGT</td>\n",
       "      <td>AAAAGT</td>\n",
       "      <td>169987319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MU64982903</td>\n",
       "      <td>TT</td>\n",
       "      <td>TCTT</td>\n",
       "      <td>169989944</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MU130883</td>\n",
       "      <td>G</td>\n",
       "      <td>ATG</td>\n",
       "      <td>169997024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MU64367222</td>\n",
       "      <td>A</td>\n",
       "      <td>GGA</td>\n",
       "      <td>170005227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>MU64018271</td>\n",
       "      <td>ATATAACAGAATT</td>\n",
       "      <td>AAATATAACAGAATT</td>\n",
       "      <td>170006654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>MU63766231</td>\n",
       "      <td>-</td>\n",
       "      <td>TTA</td>\n",
       "      <td>170018426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>MU64002203</td>\n",
       "      <td>TTAA</td>\n",
       "      <td>CTTTAA</td>\n",
       "      <td>170034676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>MU63446893</td>\n",
       "      <td>A</td>\n",
       "      <td>GGA</td>\n",
       "      <td>170034775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>MU64925025</td>\n",
       "      <td>A</td>\n",
       "      <td>AGA</td>\n",
       "      <td>170041878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>MU43769937</td>\n",
       "      <td>T</td>\n",
       "      <td>CCT</td>\n",
       "      <td>170058056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>MU64803253</td>\n",
       "      <td>T</td>\n",
       "      <td>GGT</td>\n",
       "      <td>17007427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>MU63430793</td>\n",
       "      <td>-</td>\n",
       "      <td>ACT</td>\n",
       "      <td>170085411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>MU64367227</td>\n",
       "      <td>T</td>\n",
       "      <td>TGT</td>\n",
       "      <td>170091923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>MU64011232</td>\n",
       "      <td>A</td>\n",
       "      <td>GCA</td>\n",
       "      <td>170115539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>MU64997386</td>\n",
       "      <td>-</td>\n",
       "      <td>AGA</td>\n",
       "      <td>170116021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>MU64350170</td>\n",
       "      <td>T</td>\n",
       "      <td>AAT</td>\n",
       "      <td>17011629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>MU64367229</td>\n",
       "      <td>G</td>\n",
       "      <td>GCG</td>\n",
       "      <td>170150578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>MU64063560</td>\n",
       "      <td>-</td>\n",
       "      <td>GGT</td>\n",
       "      <td>170157256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>MU63410569</td>\n",
       "      <td>TATGCT</td>\n",
       "      <td>ACTATGCT</td>\n",
       "      <td>17017047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>MU64631113</td>\n",
       "      <td>A</td>\n",
       "      <td>TGA</td>\n",
       "      <td>170179286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>MU63812497</td>\n",
       "      <td>T</td>\n",
       "      <td>CAT</td>\n",
       "      <td>170179697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>MU63631765</td>\n",
       "      <td>-</td>\n",
       "      <td>TCT</td>\n",
       "      <td>170193904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>MU65005483</td>\n",
       "      <td>AAGT</td>\n",
       "      <td>AGAAGT</td>\n",
       "      <td>170202505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>MU64367232</td>\n",
       "      <td>G</td>\n",
       "      <td>CTG</td>\n",
       "      <td>170208717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>MU64367249</td>\n",
       "      <td>ATT</td>\n",
       "      <td>AAATT</td>\n",
       "      <td>170420944</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>MU64367252</td>\n",
       "      <td>T</td>\n",
       "      <td>CAT</td>\n",
       "      <td>170421411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>MU63501914</td>\n",
       "      <td>T</td>\n",
       "      <td>AGT</td>\n",
       "      <td>170434087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>MU64631133</td>\n",
       "      <td>A</td>\n",
       "      <td>ACA</td>\n",
       "      <td>170434765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>MU64631137</td>\n",
       "      <td>T</td>\n",
       "      <td>AAT</td>\n",
       "      <td>170448688</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>MU63913109</td>\n",
       "      <td>-</td>\n",
       "      <td>TCT</td>\n",
       "      <td>170449121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>MU64367256</td>\n",
       "      <td>A</td>\n",
       "      <td>TTA</td>\n",
       "      <td>170455975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>MU64367259</td>\n",
       "      <td>T</td>\n",
       "      <td>TGT</td>\n",
       "      <td>170456137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>MU64808483</td>\n",
       "      <td>-</td>\n",
       "      <td>TGT</td>\n",
       "      <td>170456137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>MU64953087</td>\n",
       "      <td>-</td>\n",
       "      <td>TGT</td>\n",
       "      <td>170460175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>MU64367262</td>\n",
       "      <td>-</td>\n",
       "      <td>ATA</td>\n",
       "      <td>170480414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>MU63413244</td>\n",
       "      <td>AGGTTAAGCTCGGAATTTGTATTACTGTCTACTAGATC</td>\n",
       "      <td>CAAGGTTAAGCTCGGAATTTGTATTACTGTCTACTAGATC</td>\n",
       "      <td>170498181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>MU64367266</td>\n",
       "      <td>T</td>\n",
       "      <td>AGT</td>\n",
       "      <td>170516526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>MU64367271</td>\n",
       "      <td>-</td>\n",
       "      <td>TAT</td>\n",
       "      <td>170517286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>MU64367275</td>\n",
       "      <td>AAAC</td>\n",
       "      <td>AAAAAC</td>\n",
       "      <td>170522858</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>MU64631140</td>\n",
       "      <td>A</td>\n",
       "      <td>TTA</td>\n",
       "      <td>170530342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>MU63501917</td>\n",
       "      <td>A</td>\n",
       "      <td>TTA</td>\n",
       "      <td>170530401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>MU64044998</td>\n",
       "      <td>C</td>\n",
       "      <td>AAC</td>\n",
       "      <td>17053520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>MU64631144</td>\n",
       "      <td>A</td>\n",
       "      <td>GTA</td>\n",
       "      <td>170537755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>MU64619301</td>\n",
       "      <td>G</td>\n",
       "      <td>ATG</td>\n",
       "      <td>17054047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>MU63442611</td>\n",
       "      <td>-</td>\n",
       "      <td>TTG</td>\n",
       "      <td>170547969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>MU63594297</td>\n",
       "      <td>-</td>\n",
       "      <td>CTG</td>\n",
       "      <td>170560048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>MU64808487</td>\n",
       "      <td>A</td>\n",
       "      <td>AGA</td>\n",
       "      <td>170561722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>MU64808490</td>\n",
       "      <td>C</td>\n",
       "      <td>CTC</td>\n",
       "      <td>170576214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>MU64631147</td>\n",
       "      <td>T</td>\n",
       "      <td>TCT</td>\n",
       "      <td>170577411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>MU64128490</td>\n",
       "      <td>G</td>\n",
       "      <td>GAG</td>\n",
       "      <td>170581158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>MU64367278</td>\n",
       "      <td>T</td>\n",
       "      <td>ACT</td>\n",
       "      <td>170585253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>MU63854840</td>\n",
       "      <td>T</td>\n",
       "      <td>TAT</td>\n",
       "      <td>170601597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>MU64367283</td>\n",
       "      <td>-</td>\n",
       "      <td>TTA</td>\n",
       "      <td>170605896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>MU64348865</td>\n",
       "      <td>C</td>\n",
       "      <td>CAC</td>\n",
       "      <td>1706070</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>86 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            ID                                expected  \\\n",
       "0   MU64619292                                       T   \n",
       "1   MU38161712                                      TC   \n",
       "2   MU63854834                                       A   \n",
       "3   MU64350164                                       T   \n",
       "4   MU63418111                                       -   \n",
       "5   MU64994227                                       -   \n",
       "6   MU63952827                                    AAGT   \n",
       "7   MU64982903                                      TT   \n",
       "8     MU130883                                       G   \n",
       "9   MU64367222                                       A   \n",
       "10  MU64018271                           ATATAACAGAATT   \n",
       "11  MU63766231                                       -   \n",
       "12  MU64002203                                    TTAA   \n",
       "13  MU63446893                                       A   \n",
       "14  MU64925025                                       A   \n",
       "15  MU43769937                                       T   \n",
       "16  MU64803253                                       T   \n",
       "17  MU63430793                                       -   \n",
       "18  MU64367227                                       T   \n",
       "19  MU64011232                                       A   \n",
       "20  MU64997386                                       -   \n",
       "21  MU64350170                                       T   \n",
       "22  MU64367229                                       G   \n",
       "23  MU64063560                                       -   \n",
       "24  MU63410569                                  TATGCT   \n",
       "25  MU64631113                                       A   \n",
       "26  MU63812497                                       T   \n",
       "27  MU63631765                                       -   \n",
       "28  MU65005483                                    AAGT   \n",
       "29  MU64367232                                       G   \n",
       "..         ...                                     ...   \n",
       "56  MU64367249                                     ATT   \n",
       "57  MU64367252                                       T   \n",
       "58  MU63501914                                       T   \n",
       "59  MU64631133                                       A   \n",
       "60  MU64631137                                       T   \n",
       "61  MU63913109                                       -   \n",
       "62  MU64367256                                       A   \n",
       "63  MU64367259                                       T   \n",
       "64  MU64808483                                       -   \n",
       "65  MU64953087                                       -   \n",
       "66  MU64367262                                       -   \n",
       "67  MU63413244  AGGTTAAGCTCGGAATTTGTATTACTGTCTACTAGATC   \n",
       "68  MU64367266                                       T   \n",
       "69  MU64367271                                       -   \n",
       "70  MU64367275                                    AAAC   \n",
       "71  MU64631140                                       A   \n",
       "72  MU63501917                                       A   \n",
       "73  MU64044998                                       C   \n",
       "74  MU64631144                                       A   \n",
       "75  MU64619301                                       G   \n",
       "76  MU63442611                                       -   \n",
       "77  MU63594297                                       -   \n",
       "78  MU64808487                                       A   \n",
       "79  MU64808490                                       C   \n",
       "80  MU64631147                                       T   \n",
       "81  MU64128490                                       G   \n",
       "82  MU64367278                                       T   \n",
       "83  MU63854840                                       T   \n",
       "84  MU64367283                                       -   \n",
       "85  MU64348865                                       C   \n",
       "\n",
       "                                         got        pos  \n",
       "0                                        TAT   16992349  \n",
       "1                                       CATC  169942588  \n",
       "2                                        GGA  169967250  \n",
       "3                                        TAT   16996995  \n",
       "4                                        AGA  169982032  \n",
       "5                                        TTA  169984697  \n",
       "6                                     AAAAGT  169987319  \n",
       "7                                       TCTT  169989944  \n",
       "8                                        ATG  169997024  \n",
       "9                                        GGA  170005227  \n",
       "10                           AAATATAACAGAATT  170006654  \n",
       "11                                       TTA  170018426  \n",
       "12                                    CTTTAA  170034676  \n",
       "13                                       GGA  170034775  \n",
       "14                                       AGA  170041878  \n",
       "15                                       CCT  170058056  \n",
       "16                                       GGT   17007427  \n",
       "17                                       ACT  170085411  \n",
       "18                                       TGT  170091923  \n",
       "19                                       GCA  170115539  \n",
       "20                                       AGA  170116021  \n",
       "21                                       AAT   17011629  \n",
       "22                                       GCG  170150578  \n",
       "23                                       GGT  170157256  \n",
       "24                                  ACTATGCT   17017047  \n",
       "25                                       TGA  170179286  \n",
       "26                                       CAT  170179697  \n",
       "27                                       TCT  170193904  \n",
       "28                                    AGAAGT  170202505  \n",
       "29                                       CTG  170208717  \n",
       "..                                       ...        ...  \n",
       "56                                     AAATT  170420944  \n",
       "57                                       CAT  170421411  \n",
       "58                                       AGT  170434087  \n",
       "59                                       ACA  170434765  \n",
       "60                                       AAT  170448688  \n",
       "61                                       TCT  170449121  \n",
       "62                                       TTA  170455975  \n",
       "63                                       TGT  170456137  \n",
       "64                                       TGT  170456137  \n",
       "65                                       TGT  170460175  \n",
       "66                                       ATA  170480414  \n",
       "67  CAAGGTTAAGCTCGGAATTTGTATTACTGTCTACTAGATC  170498181  \n",
       "68                                       AGT  170516526  \n",
       "69                                       TAT  170517286  \n",
       "70                                    AAAAAC  170522858  \n",
       "71                                       TTA  170530342  \n",
       "72                                       TTA  170530401  \n",
       "73                                       AAC   17053520  \n",
       "74                                       GTA  170537755  \n",
       "75                                       ATG   17054047  \n",
       "76                                       TTG  170547969  \n",
       "77                                       CTG  170560048  \n",
       "78                                       AGA  170561722  \n",
       "79                                       CTC  170576214  \n",
       "80                                       TCT  170577411  \n",
       "81                                       GAG  170581158  \n",
       "82                                       ACT  170585253  \n",
       "83                                       TAT  170601597  \n",
       "84                                       TTA  170605896  \n",
       "85                                       CAC    1706070  \n",
       "\n",
       "[86 rows x 4 columns]"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Convert the invalidated mutations to a pandas dataset for convenience\n",
    "invalid_grch37_df = pd.DataFrame(invalid_grch37)\n",
    "invalid_grch37_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ID          MU38161712\n",
       "expected            TC\n",
       "got               CATC\n",
       "pos          169942588\n",
       "Name: 1, dtype: object"
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Get an entry that is not already in the invalidated\n",
    "m_not_valid_2 = invalid_grch37_df.iloc[1]\n",
    "m_not_valid_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\t169942588\tMU38161712\tATC\tA\t.\t.\tCONSEQUENCE=DHRS9|ENSG00000073737|+|DHRS9-201|ENST00000327239||intron_variant||,DHRS9|ENSG00000073737|+|DHRS9-001|ENST00000357546||intron_variant||,DHRS9|ENSG00000073737|+|DHRS9-202|ENST00000412271||intron_variant||,DHRS9|ENSG00000073737|+|DHRS9-008|ENST00000421653||intron_variant||,DHRS9|ENSG00000073737|+|DHRS9-005|ENST00000428522||intron_variant||,DHRS9|ENSG00000073737|+|DHRS9-203|ENST00000432060||intron_variant||,DHRS9|ENSG00000073737|+|DHRS9-004|ENST00000436483||intron_variant||,DHRS9|ENSG00000073737|+|DHRS9-006|ENST00000450153||downstream_gene_variant||,DHRS9|ENSG00000073737|+|DHRS9-002|ENST00000602501||intron_variant||;OCCURRENCE=ESAD-UK|1|203|0.00493,BRCA-EU|2|560|0.00357;affected_donors=3;mutation=TC>-;project_count=2;tested_donors=10638\n",
      "2\t169942588\tMU38161712\tATC\tA\t.\t.\tCONSEQUENCE=DHRS9|ENSG00000073737|+|DHRS9-201|ENST00000327239||intron_variant||,DHRS9|ENSG00000073737|+|DHRS9-001|ENST00000357546||intron_variant||,DHRS9|ENSG00000073737|+|DHRS9-202|ENST00000412271||intron_variant||,DHRS9|ENSG00000073737|+|DHRS9-008|ENST00000421653||intron_variant||,DHRS9|ENSG00000073737|+|DHRS9-005|ENST00000428522||intron_variant||,DHRS9|ENSG00000073737|+|DHRS9-203|ENST00000432060||intron_variant||,DHRS9|ENSG00000073737|+|DHRS9-004|ENST00000436483||intron_variant||,DHRS9|ENSG00000073737|+|DHRS9-006|ENST00000450153||downstream_gene_variant||,DHRS9|ENSG00000073737|+|DHRS9-002|ENST00000602501||intron_variant||;OCCURRENCE=ESAD-UK|1|253|0.00395,BRCA-EU|2|569|0.00351;affected_donors=3;mutation=TC>-;project_count=2;studies=.;tested_donors=12198\n",
      "MU38161712\tTC>-\tchr2:169942588\tchr2:169086078\tINTRONIC\tENSG00000073737(DHRS9)\tENSG00000073737(DHRS9):intron_variant,ENSG00000073737(DHRS9):downstream_gene_variant\tBRCA-EU\t\n"
     ]
    }
   ],
   "source": [
    "#CHROM    POS    ID    REF    ALT    QUAL    FILTER    INFO\n",
    "\n",
    "## From data release 22\n",
    "_ = ! grep MU38161712 ../data/data-release-22/ssm.vcf\n",
    "print(_[0])\n",
    "\n",
    "## From data release 25\n",
    "_ = ! grep MU38161712 ../data/data-release-25/ssm.vcf\n",
    "print(_[0])\n",
    "\n",
    "## From the analysis file\n",
    "_ = ! grep MU38161712 ~/results/all.BRCA-EU.mutations-context.tsv\n",
    "print(_[0])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
